// @motion-canvas/core/lib/types/Color.js
import { Color, mix } from 'chroma-js';
import { SignalContext } from '../signals';
// iife prevents tree shaking from stripping our methods.
const ExtendedColor = (() => {
    Color.symbol = Color.prototype.symbol = Symbol.for('@motion-canvas/core/types/Color');
    Color.lerp = Color.prototype.lerp = (from, to, value, colorSpace = 'lch') => {
        if (typeof from === 'string') {
            from = new Color(from);
        }
        if (typeof to === 'string') {
            to = new Color(to);
        }
        const fromIsColor = from instanceof Color;
        const toIsColor = to instanceof Color;
        if (!fromIsColor) {
            from = toIsColor ? to.alpha(0) : new Color('rgba(0, 0, 0, 0)');
        }
        if (!toIsColor) {
            to = fromIsColor
                ? from.alpha(0)
                : new Color('rgba(0, 0, 0, 0)');
        }
        return mix(from, to, value, colorSpace);
    };
    Color.createLerp = Color.prototype.createLerp =
        (colorSpace) => (from, to, value) => Color.lerp(from, to, value, colorSpace);
    Color.createSignal = (initial, interpolation = Color.lerp) => {
        return new SignalContext(initial, interpolation, undefined, value => new Color(value)).toSignal();
    };
    Color.prototype.toSymbol = () => {
        return Color.symbol;
    };
    Color.prototype.toUniform = function (gl, location) {
        gl.uniform4fv(location, this.gl());
    };
    Color.prototype.serialize = function () {
        return this.css();
    };
    Color.prototype.lerp = function (to, value, colorSpace) {
        return Color.lerp(this, to, value, colorSpace);
    };
    return Color;
})();
export { ExtendedColor as Color };
