// @motion-canvas/core/lib/types/Origin.js
import { Vector2 } from './Vector';
export var Center;
(function (Center) {
    Center[Center["Vertical"] = 1] = "Vertical";
    Center[Center["Horizontal"] = 2] = "Horizontal";
})(Center || (Center = {}));
export var Direction;
(function (Direction) {
    Direction[Direction["Top"] = 4] = "Top";
    Direction[Direction["Bottom"] = 8] = "Bottom";
    Direction[Direction["Left"] = 16] = "Left";
    Direction[Direction["Right"] = 32] = "Right";
})(Direction || (Direction = {}));
export var Origin;
(function (Origin) {
    Origin[Origin["Middle"] = 3] = "Middle";
    Origin[Origin["Top"] = 5] = "Top";
    Origin[Origin["Bottom"] = 9] = "Bottom";
    Origin[Origin["Left"] = 18] = "Left";
    Origin[Origin["Right"] = 34] = "Right";
    Origin[Origin["TopLeft"] = 20] = "TopLeft";
    Origin[Origin["TopRight"] = 36] = "TopRight";
    Origin[Origin["BottomLeft"] = 24] = "BottomLeft";
    Origin[Origin["BottomRight"] = 40] = "BottomRight";
})(Origin || (Origin = {}));
export function flipOrigin(origin, axis = Center.Horizontal | Center.Vertical) {
    if (axis & Center.Vertical) {
        if (origin & Direction.Top) {
            origin = (origin & ~Direction.Top) | Direction.Bottom;
        }
        else if (origin & Direction.Bottom) {
            origin = (origin & ~Direction.Bottom) | Direction.Top;
        }
    }
    if (axis & Center.Horizontal) {
        if (origin & Direction.Left) {
            origin = (origin & ~Direction.Left) | Direction.Right;
        }
        else if (origin & Direction.Right) {
            origin = (origin & ~Direction.Right) | Direction.Left;
        }
    }
    return origin;
}
/**
 * Convert the given origin to a vector representing its offset.
 *
 * @example
 * ```ts
 * const bottomRight = originToOffset(Origin.TopRight);
 * // bottomRight = {x: 1, y: -1}
 * ```
 *
 * @param origin - The origin to convert.
 */
export function originToOffset(origin) {
    if (origin === Origin.Middle) {
        return Vector2.zero;
    }
    let x = 0;
    if (origin & Direction.Left) {
        x = -1;
    }
    else if (origin & Direction.Right) {
        x = 1;
    }
    let y = 0;
    if (origin & Direction.Top) {
        y = -1;
    }
    else if (origin & Direction.Bottom) {
        y = 1;
    }
    return new Vector2(x, y);
}
