// @motion-canvas/core/lib/types/Spacing.js
import { CompoundSignalContext } from '../signals';
import { map } from '../tweening';
export class Spacing {
    static createSignal(initial, interpolation = Spacing.lerp) {
        return new CompoundSignalContext(['top', 'right', 'bottom', 'left'], (value) => new Spacing(value), initial, interpolation).toSignal();
    }
    static lerp(from, to, value) {
        return new Spacing(map(from.top, to.top, value), map(from.right, to.right, value), map(from.bottom, to.bottom, value), map(from.left, to.left, value));
    }
    get x() {
        return this.left + this.right;
    }
    get y() {
        return this.top + this.bottom;
    }
    constructor(one = 0, two, three, four) {
        this.top = 0;
        this.right = 0;
        this.bottom = 0;
        this.left = 0;
        if (one === undefined || one === null) {
            return;
        }
        if (Array.isArray(one)) {
            four = one[3];
            three = one[2];
            two = one[1];
            one = one[0];
        }
        if (typeof one === 'number') {
            this.top = one;
            this.right = two !== undefined ? two : one;
            this.bottom = three !== undefined ? three : one;
            this.left = four !== undefined ? four : two !== undefined ? two : one;
            return;
        }
        this.top = one.top;
        this.right = one.right;
        this.bottom = one.bottom;
        this.left = one.left;
    }
    lerp(to, value) {
        return Spacing.lerp(this, to, value);
    }
    scale(value) {
        return new Spacing(this.top * value, this.right * value, this.bottom * value, this.left * value);
    }
    addScalar(value) {
        return new Spacing(this.top + value, this.right + value, this.bottom + value, this.left + value);
    }
    toSymbol() {
        return Spacing.symbol;
    }
    toString() {
        return `Spacing(${this.top}, ${this.right}, ${this.bottom}, ${this.left})`;
    }
    toUniform(gl, location) {
        gl.uniform4f(location, this.top, this.right, this.bottom, this.left);
    }
    serialize() {
        return {
            top: this.top,
            right: this.right,
            bottom: this.bottom,
            left: this.left,
        };
    }
}
Spacing.symbol = Symbol.for('@motion-canvas/core/types/Spacing');
