// @motion-canvas/core/lib/media/AudioManager.js
import { ValueDispatcher } from '../events';
import { useLogger } from '../utils';
export class AudioManager {
    get onDataChanged() {
        return this.data.subscribable;
    }
    constructor(logger) {
        this.logger = logger;
        this.data = new ValueDispatcher(null);
        this.context = new AudioContext();
        this.audioElement = new Audio();
        this.source = null;
        this.error = false;
        this.abortController = null;
        this.offset = 0;
        if (import.meta.hot) {
            import.meta.hot.on('motion-canvas:assets', ({ urls }) => {
                if (this.source && urls.includes(this.source)) {
                    this.setSource(this.source);
                }
            });
        }
    }
    getTime() {
        return this.toAbsoluteTime(this.audioElement.currentTime);
    }
    setTime(value) {
        this.audioElement.currentTime = this.toRelativeTime(value);
    }
    setOffset(value) {
        this.offset = value;
    }
    setMuted(isMuted) {
        this.audioElement.muted = isMuted;
    }
    setVolume(volume) {
        this.audioElement.volume = volume;
    }
    setSource(src) {
        this.source = src;
        this.audioElement.src = src;
        this.abortController?.abort();
        this.abortController = new AbortController();
        this.loadData(this.abortController.signal).catch(e => {
            if (e.name !== 'AbortError') {
                this.logger.error(e);
            }
        });
    }
    isInRange(time) {
        return time >= this.offset && time < this.audioElement.duration;
    }
    toRelativeTime(time) {
        return Math.max(0, time - this.offset);
    }
    toAbsoluteTime(time) {
        return time + this.offset;
    }
    isReady() {
        return this.source && !this.error;
    }
    /**
     * Pause/resume the audio.
     *
     * @param isPaused - Whether the audio should be paused or resumed.
     *
     * @returns `true` if the audio successfully started playing.
     */
    async setPaused(isPaused) {
        if (this.source && this.audioElement.paused !== isPaused) {
            if (isPaused) {
                this.audioElement.pause();
            }
            else {
                try {
                    await this.audioElement.play();
                    this.error = false;
                    return true;
                }
                catch (e) {
                    if (!this.error) {
                        useLogger().error(e);
                    }
                    this.error = true;
                }
            }
        }
        return false;
    }
    async loadData(signal) {
        this.data.current = null;
        if (!this.source) {
            return;
        }
        const response = await fetch(this.source, { signal });
        const rawBuffer = await response.arrayBuffer();
        if (signal.aborted)
            return;
        let audioBuffer;
        try {
            audioBuffer = await this.decodeAudioData(rawBuffer);
        }
        catch (e) {
            return;
        }
        if (signal.aborted)
            return;
        const sampleSize = 256;
        const samples = ~~(audioBuffer.length / sampleSize);
        const peaks = [];
        let absoluteMax = 0;
        for (let channelId = 0; channelId < audioBuffer.numberOfChannels; channelId++) {
            const channel = audioBuffer.getChannelData(channelId);
            for (let i = 0; i < samples; i++) {
                const start = ~~(i * sampleSize);
                const end = ~~(start + sampleSize);
                let min = channel[start];
                let max = min;
                for (let j = start; j < end; j++) {
                    const value = channel[j];
                    if (value > max) {
                        max = value;
                    }
                    if (value < min) {
                        min = value;
                    }
                }
                if (channelId === 0 || max > peaks[i * 2]) {
                    peaks[i * 2] = max;
                }
                if (channelId === 0 || min < peaks[i * 2 + 1]) {
                    peaks[i * 2 + 1] = min;
                }
                if (max > absoluteMax) {
                    absoluteMax = max;
                }
                if (Math.abs(min) > absoluteMax) {
                    absoluteMax = Math.abs(min);
                }
            }
        }
        this.data.current = {
            peaks,
            absoluteMax,
            length: samples,
            sampleRate: (audioBuffer.sampleRate / sampleSize) * 2,
        };
    }
    decodeAudioData(buffer) {
        return new Promise((resolve, reject) => this.context.decodeAudioData(buffer, resolve, reject).catch(reject));
    }
}
