// @motion-canvas/core/lib/signals/Vector2SignalContext.js
import { CompoundSignalContext } from './CompoundSignalContext';
export class Vector2SignalContext extends CompoundSignalContext {
    constructor(entries, parser, initial, interpolation, owner = undefined, extensions = {}) {
        super(entries, parser, initial, interpolation, owner, extensions);
        Object.defineProperty(this.invokable, 'edit', {
            value: this.edit.bind(this),
        });
        Object.defineProperty(this.invokable, 'mul', {
            value: this.mul.bind(this),
        });
        Object.defineProperty(this.invokable, 'div', {
            value: this.div.bind(this),
        });
        Object.defineProperty(this.invokable, 'add', {
            value: this.add.bind(this),
        });
        Object.defineProperty(this.invokable, 'sub', {
            value: this.sub.bind(this),
        });
        Object.defineProperty(this.invokable, 'dot', {
            value: this.dot.bind(this),
        });
        Object.defineProperty(this.invokable, 'cross', {
            value: this.cross.bind(this),
        });
        Object.defineProperty(this.invokable, 'mod', {
            value: this.mod.bind(this),
        });
    }
    toSignal() {
        return this.invokable;
    }
    edit(callback, duration, timingFunction, interpolationFunction) {
        const newValue = callback(this.get());
        return this.invoke(newValue, duration, timingFunction, interpolationFunction);
    }
    mul(value, duration, timingFunction, interpolationFunction) {
        const callback = (current) => current.mul(value);
        if (duration === undefined)
            return this.edit(callback);
        return this.edit(callback, duration, timingFunction, interpolationFunction);
    }
    div(value, duration, timingFunction, interpolationFunction) {
        const callback = (current) => current.div(value);
        if (duration === undefined)
            return this.edit(callback);
        return this.edit(callback, duration, timingFunction, interpolationFunction);
    }
    add(value, duration, timingFunction, interpolationFunction) {
        const callback = (current) => current.add(value);
        if (duration === undefined)
            return this.edit(callback);
        return this.edit(callback, duration, timingFunction, interpolationFunction);
    }
    sub(value, duration, timingFunction, interpolationFunction) {
        const callback = (current) => current.sub(value);
        if (duration === undefined)
            return this.edit(callback);
        return this.edit(callback, duration, timingFunction, interpolationFunction);
    }
    dot(value, duration, timingFunction, interpolationFunction) {
        const callback = (current) => current.dot(value);
        if (duration === undefined)
            return this.edit(callback);
        return this.edit(callback, duration, timingFunction, interpolationFunction);
    }
    cross(value, duration, timingFunction, interpolationFunction) {
        const callback = (current) => current.cross(value);
        if (duration === undefined)
            return this.edit(callback);
        return this.edit(callback, duration, timingFunction, interpolationFunction);
    }
    mod(value, duration, timingFunction, interpolationFunction) {
        const callback = (current) => current.mod(value);
        if (duration === undefined)
            return this.edit(callback);
        return this.edit(callback, duration, timingFunction, interpolationFunction);
    }
}
