// @motion-canvas/2d/lib/curves/ArcSegment.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { BBox, DEG2RAD, Matrix2D, Vector2, lazy } from '@motion-canvas/core';
import { View2D } from '../components/View2D';
import { Segment } from './Segment';
export class ArcSegment extends Segment {
    constructor(startPoint, radius, xAxisRotationDegree, largeArcFlag, sweepFlag, endPoint) {
        super();
        this.startPoint = startPoint;
        this.radius = radius;
        this.xAxisRotationDegree = xAxisRotationDegree;
        this.largeArcFlag = largeArcFlag;
        this.sweepFlag = sweepFlag;
        this.endPoint = endPoint;
        this.xAxisRotation = this.xAxisRotationDegree * DEG2RAD;
        this.radius = new Vector2(Math.abs(radius.x), Math.abs(radius.y));
        const pAccent = startPoint
            .sub(endPoint)
            .div(2)
            .transform(Matrix2D.fromRotation(-xAxisRotationDegree).domMatrix);
        const L = (pAccent.x * pAccent.x) / (radius.x * radius.x) +
            (pAccent.y * pAccent.y) / (radius.y * radius.y);
        if (L > 1) {
            const Lsqrt = Math.sqrt(L);
            radius.x = Lsqrt * radius.x;
            radius.y = Lsqrt * radius.y;
        }
        const cAccent = new Vector2(radius.ctg * pAccent.y, radius.perpendicular.ctg * pAccent.x).scale(Math.sqrt(1 /
            ((pAccent.x * pAccent.x) / (radius.x * radius.x) +
                (pAccent.y * pAccent.y) / (radius.y * radius.y)) -
            1) * (largeArcFlag === sweepFlag ? -1 : 1));
        this.xAxisRotationMatrix =
            Matrix2D.fromRotation(xAxisRotationDegree).domMatrix;
        this.center = cAccent
            .transform(this.xAxisRotationMatrix)
            .add(startPoint.add(endPoint).div(2));
        const q = pAccent.sub(cAccent).div(radius);
        const s = pAccent.scale(-1).sub(cAccent).div(radius);
        this.startAngle = q.radians;
        this.deltaAngle = Vector2.angleBetween(q, s) % (Math.PI * 2);
        if (this.sweepFlag === 0 && this.deltaAngle > 0) {
            this.deltaAngle -= Math.PI * 2;
        }
        if (this.sweepFlag === 1 && this.deltaAngle < 0) {
            this.deltaAngle += Math.PI * 2;
        }
        ArcSegment.el.setAttribute('d', `M ${this.startPoint.x} ${this.startPoint.y} A ${this.radius.x} ${this.radius.y} ${this.xAxisRotationDegree} ${this.largeArcFlag} ${this.sweepFlag} ${this.endPoint.x} ${this.endPoint.y}`);
        this.length = ArcSegment.el.getTotalLength();
        const bbox = new BBox(ArcSegment.el.getBBox());
        this.points = [bbox.topLeft, bbox.bottomRight];
    }
    getAnglePosition(angle) {
        return this.radius
            .mul(Vector2.fromRadians(angle))
            .transform(this.xAxisRotationMatrix)
            .add(this.center);
    }
    getAngleDerivative(angle) {
        return new Vector2(-this.radius.x * Math.sin(angle), this.radius.y * Math.cos(angle)).transform(this.xAxisRotationMatrix);
    }
    draw(context, start, end, move) {
        const startAngle = this.startAngle + this.deltaAngle * start;
        const endAngle = this.startAngle + this.deltaAngle * end;
        const startPos = this.getPoint(start);
        const endPos = this.getPoint(end);
        if (move)
            context.moveTo(startPos.position.x, startPos.position.y);
        context.ellipse(this.center.x, this.center.y, this.radius.x, this.radius.y, this.xAxisRotation, startAngle, endAngle, this.sweepFlag === 0);
        return [startPos, endPos];
    }
    getPoint(distance) {
        const angle = this.startAngle + distance * this.deltaAngle;
        const tangent = this.getAngleDerivative(angle).normalized;
        return {
            position: distance === 0
                ? this.startPoint
                : distance === 1
                    ? this.endPoint
                    : this.getAnglePosition(angle),
            tangent,
            normal: tangent.perpendicular,
        };
    }
    get arcLength() {
        return this.length;
    }
}
__decorate([
    lazy(() => {
        const root = document.createElementNS('http://www.w3.org/2000/svg', 'svg');
        const el = document.createElementNS('http://www.w3.org/2000/svg', 'path');
        root.appendChild(el);
        View2D.shadowRoot.appendChild(root);
        return el;
    })
], ArcSegment, "el", void 0);
