// @motion-canvas/2d/lib/curves/Polynomial2D.js
import { BBox, Vector2 } from '@motion-canvas/core';
import { Polynomial } from './Polynomial';
export class Polynomial2D {
    constructor(c0, c1, c2, c3) {
        this.c0 = c0;
        this.c1 = c1;
        this.c2 = c2;
        this.c3 = c3;
        if (c0 instanceof Polynomial) {
            this.x = c0;
            this.y = c1;
        }
        else if (c3 !== undefined) {
            this.x = new Polynomial(c0.x, c1.x, c2.x, c3.x);
            this.y = new Polynomial(c0.y, c1.y, c2.y, c3.y);
        }
        else {
            this.x = new Polynomial(c0.x, c1.x, c2.x);
            this.y = new Polynomial(c0.y, c1.y, c2.y);
        }
    }
    eval(t, derivative = 0) {
        return new Vector2(this.x.differentiate(derivative).eval(t), this.y.differentiate(derivative).eval(t));
    }
    split(u) {
        const [xPre, xPost] = this.x.split(u);
        const [yPre, yPost] = this.y.split(u);
        return [new Polynomial2D(xPre, yPre), new Polynomial2D(xPost, yPost)];
    }
    differentiate(n = 1) {
        return new Polynomial2D(this.x.differentiate(n), this.y.differentiate(n));
    }
    evalDerivative(t) {
        return this.differentiate().eval(t);
    }
    /**
     * Calculate the tight axis-aligned bounds of the curve in the unit interval.
     */
    getBounds() {
        const rangeX = this.x.outputRange01();
        const rangeY = this.y.outputRange01();
        return BBox.fromPoints(new Vector2(Math.min(...rangeX), Math.max(...rangeY)), new Vector2(Math.max(...rangeX), Math.min(...rangeY)));
    }
}
