// @motion-canvas/2d/lib/curves/PolynomialSegment.js
import { moveTo } from '../utils';
import { Segment } from './Segment';
import { UniformPolynomialCurveSampler } from './UniformPolynomialCurveSampler';
export class PolynomialSegment extends Segment {
    get arcLength() {
        return this.length;
    }
    constructor(curve, length) {
        super();
        this.curve = curve;
        this.length = length;
        this.pointSampler = new UniformPolynomialCurveSampler(this);
    }
    getBBox() {
        return this.curve.getBounds();
    }
    /**
     * Evaluate the polynomial at the given t value.
     *
     * @param t - The t value at which to evaluate the curve.
     */
    eval(t) {
        const tangent = this.tangent(t);
        return {
            position: this.curve.eval(t),
            tangent,
            normal: tangent.perpendicular,
        };
    }
    getPoint(distance) {
        const closestPoint = this.pointSampler.pointAtDistance(this.arcLength * distance);
        return {
            position: closestPoint.position,
            tangent: closestPoint.tangent,
            normal: closestPoint.tangent.perpendicular,
        };
    }
    transformPoints(matrix) {
        return this.points.map(point => point.transformAsPoint(matrix));
    }
    /**
     * Return the tangent of the point that sits at the provided t value on the
     * curve.
     *
     * @param t - The t value at which to evaluate the curve.
     */
    tangent(t) {
        return this.curve.evalDerivative(t).normalized;
    }
    draw(context, start = 0, end = 1, move = true) {
        let curve = null;
        let startT = start;
        let endT = end;
        let points = this.points;
        if (start !== 0 || end !== 1) {
            const startDistance = this.length * start;
            const endDistance = this.length * end;
            startT = this.pointSampler.distanceToT(startDistance);
            endT = this.pointSampler.distanceToT(endDistance);
            const relativeEndT = (endT - startT) / (1 - startT);
            const [, startSegment] = this.split(startT);
            [curve] = startSegment.split(relativeEndT);
            points = curve.points;
        }
        if (move) {
            moveTo(context, points[0]);
        }
        (curve ?? this).doDraw(context);
        const startTangent = this.tangent(startT);
        const endTangent = this.tangent(endT);
        return [
            {
                position: points[0],
                tangent: startTangent,
                normal: startTangent.perpendicular,
            },
            {
                position: points.at(-1),
                tangent: endTangent,
                normal: endTangent.perpendicular,
            },
        ];
    }
}
