// @motion-canvas/core/lib/threading/Thread.js
import { EventDispatcher } from '../events';
import { noop } from '../flow';
import { createSignal } from '../signals';
import { endThread, startThread, useLogger } from '../utils';
import { isThreadGenerator } from './ThreadGenerator';
import { getTaskName, setTaskName } from './names';
/**
 * A class representing an individual thread.
 *
 * @remarks
 * Thread is a wrapper for a generator that can be executed concurrently.
 *
 * Aside from the main thread, all threads need to have a parent.
 * If a parent finishes execution, all of its child threads are terminated.
 */
export class Thread {
    get onDeferred() {
        return this.deferred.subscribable;
    }
    /**
     * The fixed time of this thread.
     *
     * @remarks
     * Fixed time is a multiple of the frame duration. It can be used to account
     * for the difference between this thread's {@link time} and the time of the
     * current animation frame.
     */
    get fixed() {
        return this.fixedTime;
    }
    /**
     * Check if this thread or any of its ancestors has been canceled.
     */
    get canceled() {
        return this.isCanceled || (this.parent?.canceled ?? false);
    }
    get paused() {
        return this.isPaused || (this.parent?.paused ?? false);
    }
    get root() {
        return this.parent?.root ?? this;
    }
    constructor(
    /**
     * The generator wrapped by this thread.
     */
    runner) {
        this.runner = runner;
        this.deferred = new EventDispatcher();
        this.children = [];
        /**
         * The current time of this thread.
         *
         * @remarks
         * Used by {@link flow.waitFor} and other time-based functions to properly
         * support durations shorter than one frame.
         */
        this.time = createSignal(0);
        this.parent = null;
        this.isCanceled = false;
        this.isPaused = false;
        this.fixedTime = 0;
        this.queue = [];
        if (this.runner.task) {
            useLogger().error({
                message: `The generator "${getTaskName(this.runner)}" is already being executed by another thread.`,
                remarks: "<p>This usually happens when you mistakenly reuse a generator that is already\nrunning.</p>\n<p>For example, using <code>yield</code> here will run the opacity generator concurrently and\nstore it in the <code>task</code> variable (in case you want to cancel or await it later):</p>\n<pre class=\"\"><code class=\"language-ts\"><span class=\"hljs-keyword\">const</span> task = <span class=\"hljs-keyword\">yield</span> <span class=\"hljs-title function_\">rect</span>().<span class=\"hljs-title function_\">opacity</span>(<span class=\"hljs-number\">1</span>, <span class=\"hljs-number\">1</span>);</code></pre><p>Trying to <code>yield</code> this task again will cause the current error:</p>\n<pre class=\"\"><code class=\"language-ts\"><span class=\"hljs-keyword\">yield</span> task;</code></pre><p>Passing it to other flow functions will also cause the error:</p>\n<pre class=\"\"><code class=\"language-ts\"><span class=\"hljs-keyword\">yield</span>* <span class=\"hljs-title function_\">all</span>(task);</code></pre><p>Try to investigate your code looking for <code>yield</code> statements whose return value\nis reused in this way. Here&#39;s an example of a common mistake:</p>\n<pre class=\"wrong\"><code class=\"language-ts\"><span class=\"hljs-keyword\">yield</span>* <span class=\"hljs-title function_\">all</span>(\n  <span class=\"hljs-keyword\">yield</span> <span class=\"hljs-title function_\">rect</span>().<span class=\"hljs-title function_\">opacity</span>(<span class=\"hljs-number\">1</span>, <span class=\"hljs-number\">1</span>), \n  <span class=\"hljs-keyword\">yield</span> <span class=\"hljs-title function_\">rect</span>().<span class=\"hljs-title function_\">x</span>(<span class=\"hljs-number\">200</span>, <span class=\"hljs-number\">1</span>),\n);</code></pre><pre class=\"correct\"><code class=\"language-ts\"><span class=\"hljs-keyword\">yield</span>* <span class=\"hljs-title function_\">all</span>(\n  <span class=\"hljs-title function_\">rect</span>().<span class=\"hljs-title function_\">opacity</span>(<span class=\"hljs-number\">1</span>, <span class=\"hljs-number\">1</span>), \n  <span class=\"hljs-title function_\">rect</span>().<span class=\"hljs-title function_\">x</span>(<span class=\"hljs-number\">200</span>, <span class=\"hljs-number\">1</span>),\n);</code></pre>",
            });
            this.runner = noop();
        }
        this.runner.task = this;
    }
    /**
     * Progress the wrapped generator once.
     */
    next() {
        if (this.paused) {
            return {
                value: null,
                done: false,
            };
        }
        startThread(this);
        const result = this.runner.next(this.value);
        endThread(this);
        this.value = null;
        return result;
    }
    /**
     * Prepare the thread for the next update cycle.
     *
     * @param dt - The delta time of the next cycle.
     */
    update(dt) {
        if (!this.paused) {
            this.time(this.time() + dt);
            this.fixedTime += dt;
        }
        this.children = this.children.filter(child => !child.canceled);
    }
    spawn(child) {
        if (!isThreadGenerator(child)) {
            child = child();
        }
        this.queue.push(child);
        return child;
    }
    add(child) {
        child.parent = this;
        child.isCanceled = false;
        child.time(this.time());
        child.fixedTime = this.fixedTime;
        this.children.push(child);
        setTaskName(child.runner, `unknown ${this.children.length}`);
    }
    drain(callback) {
        this.queue.forEach(callback);
        this.queue = [];
    }
    cancel() {
        this.deferred.clear();
        this.runner.return();
        this.isCanceled = true;
        this.parent = null;
        this.drain(task => task.return());
    }
    pause(value) {
        this.isPaused = value;
    }
    runDeferred() {
        startThread(this);
        this.deferred.dispatch();
        endThread(this);
    }
}
