// @motion-canvas/core/lib/scenes/LifecycleEvents.js
import { EventDispatcher } from '../events';
import { SceneRenderEvent } from './Scene';
/**
 * Lifecycle events for {@link Scene} that are cleared on every reset.
 */
export class LifecycleEvents {
    get onBeforeRender() {
        return this.beforeRender.subscribable;
    }
    get onBeginRender() {
        return this.beginRender.subscribable;
    }
    get onFinishRender() {
        return this.finishRender.subscribable;
    }
    get onAfterRender() {
        return this.afterRender.subscribable;
    }
    constructor(scene) {
        this.scene = scene;
        this.beforeRender = new EventDispatcher();
        this.beginRender = new EventDispatcher();
        this.finishRender = new EventDispatcher();
        this.afterRender = new EventDispatcher();
        this.scene.onRenderLifecycle.subscribe(([event, ctx]) => {
            switch (event) {
                case SceneRenderEvent.BeforeRender:
                    return this.beforeRender.dispatch(ctx);
                case SceneRenderEvent.BeginRender:
                    return this.beginRender.dispatch(ctx);
                case SceneRenderEvent.FinishRender:
                    return this.finishRender.dispatch(ctx);
                case SceneRenderEvent.AfterRender:
                    return this.afterRender.dispatch(ctx);
            }
        });
        this.scene.onReset.subscribe(() => {
            this.beforeRender.clear();
            this.beginRender.clear();
            this.finishRender.clear();
            this.afterRender.clear();
        });
    }
}
