// @motion-canvas/core/lib/scenes/Slides.js
import { PlaybackState } from '../app/PlaybackManager';
import { ValueDispatcher } from '../events';
export class Slides {
    get onChanged() {
        return this.slides.subscribable;
    }
    constructor(scene) {
        this.scene = scene;
        this.slides = new ValueDispatcher([]);
        this.lookup = new Map();
        this.collisionLookup = new Set();
        this.current = null;
        this.canResume = false;
        this.waitsForId = null;
        this.targetId = null;
        this.handleReload = () => {
            this.lookup.clear();
            this.collisionLookup.clear();
            this.current = null;
            this.waitsForId = null;
            this.targetId = null;
        };
        this.handleReset = () => {
            this.collisionLookup.clear();
            this.current = null;
            this.waitsForId = null;
        };
        this.handleRecalculated = () => {
            this.slides.current = [...this.lookup.values()];
        };
        this.scene.onReloaded.subscribe(this.handleReload);
        this.scene.onReset.subscribe(this.handleReset);
        this.scene.onRecalculated.subscribe(this.handleRecalculated);
    }
    setTarget(target) {
        this.targetId = target;
    }
    resume() {
        this.canResume = true;
    }
    isWaitingFor(slide) {
        return this.waitsForId === slide;
    }
    isWaiting() {
        return this.waitsForId !== null;
    }
    didHappen(slide) {
        if (this.current === null) {
            return false;
        }
        for (const key of this.lookup.keys()) {
            if (key === slide) {
                return true;
            }
            if (key === this.current?.id) {
                return false;
            }
        }
        return false;
    }
    getCurrent() {
        return this.current;
    }
    register(name, initialTime) {
        if (this.waitsForId !== null) {
            throw new Error(`The animation already waits for a slide: ${this.waitsForId}.`);
        }
        const id = this.toId(name);
        if (this.scene.playback.state !== PlaybackState.Presenting) {
            if (!this.lookup.has(id)) {
                this.lookup.set(id, {
                    id,
                    name,
                    time: initialTime,
                    scene: this.scene,
                    stack: new Error().stack,
                });
            }
            if (this.collisionLookup.has(name)) {
                this.scene.logger.warn({
                    message: `A slide named "${name}" already exists.`,
                    stack: new Error().stack,
                });
            }
            else {
                this.collisionLookup.add(name);
            }
        }
        this.waitsForId = id;
        this.current = this.lookup.get(id) ?? null;
        this.canResume = false;
    }
    shouldWait(name) {
        const id = this.toId(name);
        if (this.waitsForId !== id) {
            throw new Error(`The animation waits for a different slide: ${this.waitsForId}.`);
        }
        const data = this.lookup.get(id);
        if (!data) {
            throw new Error(`Could not find the "${name}" slide.`);
        }
        let canResume = this.canResume;
        if (this.scene.playback.state !== PlaybackState.Presenting) {
            canResume = id !== this.targetId;
        }
        if (canResume) {
            this.waitsForId = null;
        }
        return !canResume;
    }
    toId(name) {
        return `${this.scene.name}:${name}`;
    }
}
