// @motion-canvas/core/lib/app/Stage.js
import { unwrap } from '../signals';
import { Vector2 } from '../types';
import { getContext } from '../utils';
/**
 * Manages canvases on which an animation can be displayed.
 */
export class Stage {
    get canvasSize() {
        return this.size.scale(this.resolutionScale);
    }
    constructor() {
        // TODO Consider adding pooling for canvases.
        this.background = null;
        this.resolutionScale = 1;
        this.colorSpace = 'srgb';
        this.size = Vector2.zero;
        this.finalBuffer = document.createElement('canvas');
        this.currentBuffer = document.createElement('canvas');
        this.previousBuffer = document.createElement('canvas');
        const colorSpace = this.colorSpace;
        this.context = getContext({ colorSpace }, this.finalBuffer);
        this.currentContext = getContext({ colorSpace }, this.currentBuffer);
        this.previousContext = getContext({ colorSpace }, this.previousBuffer);
    }
    configure({ colorSpace = this.colorSpace, size = this.size, resolutionScale = this.resolutionScale, background = this.background, }) {
        if (colorSpace !== this.colorSpace) {
            this.colorSpace = colorSpace;
            this.context = getContext({ colorSpace }, this.finalBuffer);
            this.currentContext = getContext({ colorSpace }, this.currentBuffer);
            this.previousContext = getContext({ colorSpace }, this.previousBuffer);
        }
        if (!size.exactlyEquals(this.size) ||
            resolutionScale !== this.resolutionScale) {
            this.resolutionScale = resolutionScale;
            this.size = size;
            this.resizeCanvas(this.context);
            this.resizeCanvas(this.currentContext);
            this.resizeCanvas(this.previousContext);
        }
        this.background =
            typeof background === 'string'
                ? background
                : background?.serialize() ?? null;
    }
    async render(currentScene, previousScene) {
        const previousOnTop = previousScene
            ? unwrap(currentScene.previousOnTop)
            : false;
        if (previousScene) {
            await previousScene.render(this.previousContext);
        }
        await currentScene.render(this.currentContext);
        const size = this.canvasSize;
        this.context.clearRect(0, 0, size.width, size.height);
        if (this.background) {
            this.context.save();
            this.context.fillStyle = this.background;
            this.context.fillRect(0, 0, size.width, size.height);
            this.context.restore();
        }
        if (previousScene && !previousOnTop) {
            this.context.drawImage(this.previousBuffer, 0, 0);
        }
        this.context.drawImage(this.currentBuffer, 0, 0);
        if (previousOnTop) {
            this.context.drawImage(this.previousBuffer, 0, 0);
        }
    }
    resizeCanvas(context) {
        const size = this.canvasSize;
        context.canvas.width = size.width;
        context.canvas.height = size.height;
    }
}
