// @motion-canvas/core/lib/app/TimeEstimator.js
import { ValueDispatcher } from '../events';
import { clamp } from '../tweening';
/**
 * Calculates the estimated time remaining until a process is finished.
 */
export class TimeEstimator {
    constructor() {
        this.completion = new ValueDispatcher(0);
        this.startTimestamp = 0;
        this.lastUpdateTimestamp = 0;
        this.nextCompletion = 0;
    }
    get onCompletionChanged() {
        return this.completion.subscribable;
    }
    /**
     * Get the current time estimate.
     *
     * @param timestamp - The timestamp to calculate the estimate against.
     *                    Defaults to `performance.now()`.
     */
    estimate(timestamp = performance.now()) {
        const elapsed = timestamp - this.startTimestamp;
        const completion = this.completion.current;
        let eta = Infinity;
        if (completion >= 1) {
            eta = 0;
        }
        else if (completion > 0) {
            const lastUpdateDuration = this.lastUpdateTimestamp - this.startTimestamp;
            eta = lastUpdateDuration / completion - elapsed;
            eta = Math.max(0, eta);
        }
        else if (this.nextCompletion > 0) {
            eta = elapsed / this.nextCompletion - elapsed;
        }
        return { completion, elapsed, eta };
    }
    /**
     * Update the completion percentage.
     *
     * @param completion - The completion percentage ranging from `0` to `1`.
     * @param timestamp - A timestamp at which the process was updated.
     *                    Defaults to `performance.now()`.
     */
    update(completion, timestamp = performance.now()) {
        this.completion.current = clamp(0, 1, completion);
        this.lastUpdateTimestamp = timestamp;
    }
    /**
     * Reset the estimator.
     *
     * @param nextCompletion - If known, the completion percentage of the next
     *                         update.
     * @param timestamp - A timestamp at which the process started.
     *                    Defaults to `performance.now()`.
     */
    reset(nextCompletion = 0, timestamp = performance.now()) {
        this.startTimestamp = timestamp;
        this.lastUpdateTimestamp = timestamp;
        this.completion.current = 0;
        this.nextCompletion = nextCompletion;
    }
}
