// @motion-canvas/ffmpeg/lib/server/FFmpegBridge.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FFmpegBridge = void 0;
const FFmpegExporterServer_1 = require("./FFmpegExporterServer");
/**
 * A simple bridge between the FFmpegExporterServer and FFmpegExporterClient.
 *
 * @remarks
 * This class lets the client exporter invoke methods on the server and receive
 * responses using a simple Promise-based API.
 */
class FFmpegBridge {
    constructor(ws, config) {
        this.ws = ws;
        this.config = config;
        this.process = null;
        this.handleMessage = async ({ method, data }) => {
            if (method === 'start') {
                try {
                    this.process = new FFmpegExporterServer_1.FFmpegExporterServer(data, this.config);
                    this.respondSuccess(method, await this.process.start());
                }
                catch (e) {
                    this.respondError(method, e?.message);
                }
                return;
            }
            if (!this.process) {
                this.respondError(method, 'The exporting process has not been started.');
                return;
            }
            if (!(method in this.process)) {
                this.respondError(method, `Unknown method: "${method}".`);
                return;
            }
            try {
                this.respondSuccess(method, await this.process[method](data));
            }
            catch (e) {
                this.respondError(method, e?.message);
            }
            if (method === 'end') {
                this.process = null;
            }
        };
        ws.on('motion-canvas/ffmpeg', this.handleMessage);
    }
    respondSuccess(method, data = {}) {
        this.ws.send('motion-canvas/ffmpeg-ack', {
            status: 'success',
            method,
            data,
        });
    }
    respondError(method, message = 'Unknown error.') {
        this.ws.send('motion-canvas/ffmpeg-ack', {
            status: 'error',
            method,
            message,
        });
    }
}
exports.FFmpegBridge = FFmpegBridge;
