// @motion-canvas/2d/lib/code/CodeFragment.js
import { isCodeMetrics, measureString } from './CodeMetrics';
export function metricsToFragment(value) {
    return {
        before: value,
        after: value,
    };
}
export function parseCodeFragment(value, context, monoWidth) {
    let fragment;
    if (typeof value === 'string') {
        fragment = metricsToFragment(measureString(context, monoWidth, value));
    }
    else if (isCodeMetrics(value)) {
        fragment = metricsToFragment(value);
    }
    else {
        fragment = {
            before: typeof value.before === 'string'
                ? measureString(context, monoWidth, value.before)
                : value.before,
            after: typeof value.after === 'string'
                ? measureString(context, monoWidth, value.after)
                : value.after,
        };
    }
    return fragment;
}
/**
 * Create a code fragment that represents an insertion of code.
 *
 * @remarks
 * Can be used in conjunction with {@link code.CodeSignalHelpers.edit}.
 *
 * @param code - The code to insert.
 */
export function insert(code) {
    return {
        before: '',
        after: code,
    };
}
/**
 * Create a code fragment that represents a change from one piece of code to
 * another.
 *
 * @remarks
 * Can be used in conjunction with {@link code.CodeSignalHelpers.edit}.
 *
 * @param before - The code to change from.
 * @param after - The code to change to.
 */
export function replace(before, after) {
    return {
        before,
        after,
    };
}
/**
 * Create a code fragment that represents a removal of code.
 *
 * @remarks
 * Can be used in conjunction with {@link code.CodeSignalHelpers.edit}.
 *
 * @param code - The code to remove.
 */
export function remove(code) {
    return {
        before: code,
        after: '',
    };
}
