// @motion-canvas/2d/lib/code/CodeScope.js
import { unwrap } from '@motion-canvas/core';
import { isCodeMetrics } from './CodeMetrics';
export function CODE(strings, ...tags) {
    const result = [];
    for (let i = 0; i < strings.length; i++) {
        result.push(strings[i]);
        const tag = tags[i];
        if (tag !== undefined) {
            if (Array.isArray(tag)) {
                result.push(...tag);
            }
            else {
                result.push(tag);
            }
        }
    }
    return result;
}
export function isCodeScope(value) {
    return value?.fragments !== undefined;
}
export function parseCodeScope(value) {
    if (typeof value === 'string') {
        return {
            progress: 0,
            fragments: [value],
        };
    }
    if (Array.isArray(value)) {
        return {
            progress: 0,
            fragments: value,
        };
    }
    return value;
}
export function resolveScope(scope, isAfter) {
    let code = '';
    const after = typeof isAfter === 'boolean' ? isAfter : isAfter(scope);
    for (const wrapped of scope.fragments) {
        code += resolveCodeTag(wrapped, after, isAfter);
    }
    return code;
}
export function resolveCodeTag(wrapped, after, isAfter = after) {
    const fragment = unwrap(wrapped);
    if (typeof fragment === 'string') {
        return fragment;
    }
    else if (isCodeScope(fragment)) {
        return resolveScope(fragment, isAfter);
    }
    else if (isCodeMetrics(fragment)) {
        return fragment.content;
    }
    else if (Array.isArray(fragment)) {
        return resolveScope({
            progress: 0,
            fragments: fragment,
        }, isAfter);
    }
    else {
        return after
            ? typeof fragment.after === 'string'
                ? fragment.after
                : fragment.after.content
            : typeof fragment.before === 'string'
                ? fragment.before
                : fragment.before.content;
    }
}
