// @motion-canvas/2d/lib/code/LezerHighlighter.js
import { highlightTree } from '@lezer/highlight';
import { DefaultHighlightStyle } from './DefaultHighlightStyle';
export class LezerHighlighter {
    constructor(parser, style = DefaultHighlightStyle) {
        this.parser = parser;
        this.style = style;
        this.classLookup = new Map();
        for (const rule of this.style.module?.getRules().split('\n') ?? []) {
            const match = rule.match(LezerHighlighter.classRegex);
            if (!match) {
                continue;
            }
            const className = match[1];
            const color = match[2].trim();
            this.classLookup.set(className, color);
        }
    }
    initialize() {
        return true;
    }
    prepare(code) {
        const colorLookup = new Map();
        const tree = this.parser.parse(code);
        highlightTree(tree, this.style, (from, to, classes) => {
            const color = this.classLookup.get(classes);
            if (!color) {
                return;
            }
            const cursor = tree.cursorAt(from, 1);
            do {
                const id = this.getNodeId(cursor.node);
                colorLookup.set(id, color);
            } while (cursor.next() && cursor.to <= to);
        });
        return {
            tree,
            code,
            colorLookup,
        };
    }
    highlight(index, cache) {
        if (!cache) {
            return {
                color: null,
                skipAhead: 0,
            };
        }
        const node = cache.tree.resolveInner(index, 1);
        const id = this.getNodeId(node);
        const color = cache.colorLookup.get(id);
        if (color) {
            return {
                color,
                skipAhead: node.to - index,
            };
        }
        let skipAhead = 0;
        if (!node.firstChild) {
            skipAhead = node.to - index;
        }
        return {
            color: null,
            skipAhead,
        };
    }
    tokenize(code) {
        const tree = this.parser.parse(code);
        const cursor = tree.cursor();
        const tokens = [];
        let current = 0;
        do {
            if (!cursor.node.firstChild) {
                if (cursor.from > current) {
                    tokens.push(code.slice(current, cursor.from));
                }
                if (cursor.from < cursor.to) {
                    tokens.push(code.slice(cursor.from, cursor.to));
                }
                current = cursor.to;
            }
        } while (cursor.next());
        return tokens;
    }
    getNodeId(node) {
        return `${node.from}:${node.to}`;
    }
}
LezerHighlighter.classRegex = /\.(\S+).*color:([^;]+)/;
