// @motion-canvas/2d/lib/components/Bezier.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { BBox } from '@motion-canvas/core';
import { computed } from '../decorators';
import { arc, drawLine, drawPivot, moveTo } from '../utils';
import { Curve } from './Curve';
export class Bezier extends Curve {
    profile() {
        const segment = this.segment();
        return {
            segments: [segment],
            arcLength: segment.arcLength,
            minSin: 0,
        };
    }
    childrenBBox() {
        return BBox.fromPoints(...this.segment().points);
    }
    desiredSize() {
        return this.segment().getBBox().size;
    }
    offsetComputedLayout(box) {
        box.position = box.position.sub(this.segment().getBBox().center);
        return box;
    }
    drawOverlay(context, matrix) {
        const size = this.computedSize();
        const box = this.childrenBBox().transformCorners(matrix);
        const offset = size.mul(this.offset()).scale(0.5).transformAsPoint(matrix);
        const overlayInfo = this.overlayInfo(matrix);
        context.lineWidth = 1;
        context.strokeStyle = 'white';
        context.fillStyle = 'white';
        // Draw the curve itself first so everything else gets drawn on top of it
        context.stroke(overlayInfo.curve);
        context.fillStyle = 'white';
        context.globalAlpha = 0.5;
        context.beginPath();
        context.stroke(overlayInfo.handleLines);
        context.globalAlpha = 1;
        context.lineWidth = 2;
        // Draw start and end points
        for (const point of [overlayInfo.startPoint, overlayInfo.endPoint]) {
            moveTo(context, point);
            context.beginPath();
            arc(context, point, 4);
            context.closePath();
            context.stroke();
            context.fill();
        }
        // Draw the control points
        context.fillStyle = 'black';
        for (const point of overlayInfo.controlPoints) {
            moveTo(context, point);
            context.beginPath();
            arc(context, point, 4);
            context.closePath();
            context.fill();
            context.stroke();
        }
        // Draw the offset marker
        context.lineWidth = 1;
        context.beginPath();
        drawPivot(context, offset);
        context.stroke();
        // Draw the bounding box
        context.beginPath();
        drawLine(context, box);
        context.closePath();
        context.stroke();
    }
}
__decorate([
    computed()
], Bezier.prototype, "childrenBBox", null);
