// @motion-canvas/2d/lib/components/Camera.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { all, DEFAULT, easeInOutCubic, modify, threadable, tween, unwrap, Vector2, } from '@motion-canvas/core';
import { cloneable, signal } from '../decorators';
import { Node } from './Node';
import { Rect } from './Rect';
/**
 * A node representing an orthographic camera.
 *
 * @preview
 * ```tsx editor
 * import {Camera, Circle, makeScene2D, Rect} from '@motion-canvas/2d';
 * import {all, createRef} from '@motion-canvas/core';
 *
 * export default makeScene2D(function* (view) {
 *   const camera = createRef<Camera>();
 *   const rect = createRef<Rect>();
 *   const circle = createRef<Circle>();
 *
 *   view.add(
 *     <>
 *       <Camera ref={camera}>
 *         <Rect
 *           ref={rect}
 *           fill={'lightseagreen'}
 *           size={100}
 *           position={[100, -50]}
 *         />
 *         <Circle
 *           ref={circle}
 *           fill={'hotpink'}
 *           size={120}
 *           position={[-100, 50]}
 *         />
 *       </Camera>
 *     </>,
 *   );
 *
 *   yield* all(
 *     camera().centerOn(rect(), 3),
 *     camera().rotation(180, 3),
 *     camera().zoom(1.8, 3),
 *   );
 *   yield* camera().centerOn(circle(), 2);
 *   yield* camera().reset(1);
 * });
 * ```
 */
export class Camera extends Node {
    constructor({ children, ...props }) {
        super(props);
        if (!this.scene()) {
            this.scene(new Node({}));
        }
        if (children) {
            this.scene().add(children);
        }
    }
    getZoom() {
        return 1 / this.scale.x();
    }
    setZoom(value) {
        this.scale(modify(value, unwrapped => 1 / unwrapped));
    }
    getDefaultZoom() {
        return this.scale.x.context.getInitial();
    }
    *tweenZoom(value, duration, timingFunction, interpolationFunction) {
        const from = this.scale.x();
        yield* tween(duration, v => {
            this.zoom(1 / interpolationFunction(from, 1 / unwrap(value), timingFunction(v)));
        });
    }
    /**
     * Resets the camera's position, rotation and zoom level to their original
     * values.
     *
     * @param duration - The duration of the tween.
     * @param timingFunction - The timing function to use for the tween.
     */
    *reset(duration, timingFunction = easeInOutCubic) {
        yield* all(this.position(DEFAULT, duration, timingFunction), this.zoom(DEFAULT, duration, timingFunction), this.rotation(DEFAULT, duration, timingFunction));
    }
    *centerOn(positionOrNode, duration, timing = easeInOutCubic, interpolationFunction = Vector2.lerp) {
        const position = positionOrNode instanceof Node
            ? positionOrNode
                .absolutePosition()
                .transformAsPoint(this.scene().worldToLocal())
            : positionOrNode;
        yield* this.position(position, duration, timing, interpolationFunction);
    }
    /**
     * Makes the camera follow a path specified by the provided curve.
     *
     * @remarks
     * This will not change the orientation of the camera. To make the camera
     * orient itself along the curve, use {@link followCurveWithRotation} or
     * {@link followCurveWithRotationReverse}.
     *
     * If you want to follow the curve in reverse, use {@link followCurveReverse}.
     *
     * @param curve - The curve to follow.
     * @param duration - The duration of the tween.
     * @param timing - The timing function to use for the tween.
     */
    *followCurve(curve, duration, timing = easeInOutCubic) {
        yield* tween(duration, value => {
            const t = timing(value);
            const point = curve
                .getPointAtPercentage(t)
                .position.transformAsPoint(curve.localToWorld());
            this.position(point);
        });
    }
    /**
     * Makes the camera follow a path specified by the provided curve in reverse.
     *
     * @remarks
     * This will not change the orientation of the camera. To make the camera
     * orient itself along the curve, use {@link followCurveWithRotation} or
     * {@link followCurveWithRotationReverse}.
     *
     * If you want to follow the curve forward, use {@link followCurve}.
     *
     * @param curve - The curve to follow.
     * @param duration - The duration of the tween.
     * @param timing - The timing function to use for the tween.
     */
    *followCurveReverse(curve, duration, timing = easeInOutCubic) {
        yield* tween(duration, value => {
            const t = 1 - timing(value);
            const point = curve
                .getPointAtPercentage(t)
                .position.transformAsPoint(curve.localToWorld());
            this.position(point);
        });
    }
    /**
     * Makes the camera follow a path specified by the provided curve while
     * pointing the camera the direction of the tangent.
     *
     * @remarks
     * To make the camera follow the curve without changing its orientation, use
     * {@link followCurve} or {@link followCurveReverse}.
     *
     * If you want to follow the curve in reverse, use
     * {@link followCurveWithRotationReverse}.
     *
     * @param curve - The curve to follow.
     * @param duration - The duration of the tween.
     * @param timing - The timing function to use for the tween.
     */
    *followCurveWithRotation(curve, duration, timing = easeInOutCubic) {
        yield* tween(duration, value => {
            const t = timing(value);
            const { position, normal } = curve.getPointAtPercentage(t);
            const point = position.transformAsPoint(curve.localToWorld());
            const angle = normal.flipped.perpendicular.degrees;
            this.position(point);
            this.rotation(angle);
        });
    }
    /**
     * Makes the camera follow a path specified by the provided curve in reverse
     * while pointing the camera the direction of the tangent.
     *
     * @remarks
     * To make the camera follow the curve without changing its orientation, use
     * {@link followCurve} or {@link followCurveReverse}.
     *
     * If you want to follow the curve forward, use
     * {@link followCurveWithRotation}.
     *
     * @param curve - The curve to follow.
     * @param duration - The duration of the tween.
     * @param timing - The timing function to use for the tween.
     */
    *followCurveWithRotationReverse(curve, duration, timing = easeInOutCubic) {
        yield* tween(duration, value => {
            const t = 1 - timing(value);
            const { position, normal } = curve.getPointAtPercentage(t);
            const point = position.transformAsPoint(curve.localToWorld());
            const angle = normal.flipped.perpendicular.degrees;
            this.position(point);
            this.rotation(angle);
        });
    }
    transformContext(context) {
        const matrix = this.localToParent().inverse();
        context.transform(matrix.a, matrix.b, matrix.c, matrix.d, matrix.e, matrix.f);
    }
    hit(position) {
        const local = position.transformAsPoint(this.localToParent());
        return this.scene().hit(local);
    }
    drawChildren(context) {
        this.scene().drawChildren(context);
    }
    // eslint-disable-next-line @typescript-eslint/naming-convention
    static Stage({ children, cameraRef, scene, ...props }) {
        const camera = new Camera({ scene: scene, children });
        cameraRef?.(camera);
        return new Rect({
            clip: true,
            ...props,
            children: [camera],
        });
    }
}
__decorate([
    signal()
], Camera.prototype, "scene", void 0);
__decorate([
    cloneable(false),
    signal()
], Camera.prototype, "zoom", void 0);
__decorate([
    threadable()
], Camera.prototype, "reset", null);
__decorate([
    threadable()
], Camera.prototype, "centerOn", null);
__decorate([
    threadable()
], Camera.prototype, "followCurve", null);
__decorate([
    threadable()
], Camera.prototype, "followCurveReverse", null);
__decorate([
    threadable()
], Camera.prototype, "followCurveWithRotation", null);
__decorate([
    threadable()
], Camera.prototype, "followCurveWithRotationReverse", null);
