// @motion-canvas/2d/lib/components/CodeBlock.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var CodeBlock_1;
import { Vector2, clampRemap, createComputedAsync, createSignal, easeInOutSine, join, map, threadable, tween, useLogger, waitFor, } from '@motion-canvas/core';
import { diff, parse, ready, } from 'code-fns';
import { computed, initial, nodeName, parser, signal } from '../decorators';
import { Shape } from './Shape';
/**
 * @deprecated Use {@link Code} instead.
 */
let CodeBlock = CodeBlock_1 = class CodeBlock extends Shape {
    *tweenSelection(value, duration, timingFunction) {
        this.oldSelection = this.selection();
        this.selection(value);
        this.selectionProgress(0);
        yield* this.selectionProgress(1, duration, timingFunction);
        this.selectionProgress(null);
        this.oldSelection = null;
    }
    getLineCountOfTokenArray(tokens) {
        let count = 0;
        for (const token of tokens) {
            for (let i = 0; i < token.code.length; i++) {
                if (token.code[i] === '\n') {
                    count++;
                }
            }
        }
        if (tokens.length > 0) {
            count++;
        }
        return count;
    }
    lineCount() {
        const progress = this.codeProgress();
        if (progress !== null) {
            return Math.round(map(this.currentLineCount, this.newLineCount, progress));
        }
        return this.getLineCountOfTokenArray(this.parsed());
    }
    parsed() {
        if (!CodeBlock_1.initialized()) {
            return [];
        }
        return parse(this.code(), { codeStyle: this.theme() });
    }
    constructor({ children, ...rest }) {
        super({
            fontFamily: 'monospace',
            ...rest,
        });
        this.codeProgress = createSignal(null);
        this.selectionProgress = createSignal(null);
        this.oldSelection = null;
        this.diffed = null;
        this.currentLineCount = 0;
        this.newLineCount = 0;
        if (children) {
            this.code(children);
        }
    }
    characterSize() {
        this.requestFontUpdate();
        const context = this.cacheCanvas();
        context.save();
        this.applyStyle(context);
        context.font = this.styles.font;
        const width = context.measureText('X').width;
        context.restore();
        return new Vector2(width, parseFloat(this.styles.lineHeight));
    }
    desiredSize() {
        const custom = super.desiredSize();
        const tokensSize = this.getTokensSize(this.parsed());
        return {
            x: custom.x ?? tokensSize.x,
            y: custom.y ?? tokensSize.y,
        };
    }
    getTokensSize(tokens) {
        const size = this.characterSize();
        let maxWidth = 0;
        let height = size.height;
        let width = 0;
        for (const token of tokens) {
            for (let i = 0; i < token.code.length; i++) {
                if (token.code[i] === '\n') {
                    if (width > maxWidth) {
                        maxWidth = width;
                    }
                    width = 0;
                    height += size.height;
                }
                else {
                    width += size.width;
                }
            }
        }
        if (width > maxWidth) {
            maxWidth = width;
        }
        return { x: maxWidth, y: height };
    }
    collectAsyncResources() {
        super.collectAsyncResources();
        CodeBlock_1.initialized();
    }
    set(strings, ...rest) {
        this.code({
            language: this.language(),
            spans: strings,
            nodes: rest,
        });
    }
    /**
     * Smoothly edit the code.
     *
     * @remarks
     * This method returns a tag function that should be used together with a
     * template literal to define what to edit. Expressions can be used to either
     * {@link insert}, {@link remove}, or {@link edit} the code.
     *
     * @example
     * ```ts
     * yield* codeBlock().edit()`
     *   const ${edit('a', 'b')} = [${insert('1, 2, 3')}];${remove(`
     *   // this comment will be removed`)}
     * `;
     * ```
     *
     * @param duration - The duration of the transition.
     * @param changeSelection - When set to `true`, the selection will be modified
     *                          to highlight the newly inserted code. Setting it
     *                          to `false` leaves the selection untouched.
     *                          Providing a custom {@link CodeRange} will select
     *                          it instead.
     */
    edit(duration = 0.6, changeSelection = true) {
        function* generator(strings, ...rest) {
            const from = {
                language: this.language(),
                spans: [...strings],
                nodes: rest.map(modification => isCodeModification(modification) ? modification.from : modification),
            };
            const to = {
                language: this.language(),
                spans: [...strings],
                nodes: rest.map(modification => isCodeModification(modification) ? modification.to : modification),
            };
            this.code(from);
            if (changeSelection) {
                const task = yield this.code(to, duration);
                yield* waitFor(duration * 0.2);
                yield* this.selection([], duration * 0.3);
                const newSelection = changeSelection === true
                    ? diff(from, to)
                        .filter(token => token.morph === 'create')
                        .map(token => [
                        [token.to[1], token.to[0]],
                        [token.to[1], token.to[0] + token.code.length],
                    ])
                    : changeSelection;
                yield* this.selection(newSelection, duration * 0.3);
                yield* join(task);
            }
            else {
                yield* this.code(to, duration);
            }
        }
        return generator.bind(this);
    }
    *tweenCode(code, time, timingFunction) {
        if (typeof code === 'function')
            throw new Error();
        if (!CodeBlock_1.initialized())
            return;
        const currentParsedCode = parse(this.code(), { codeStyle: this.theme() });
        const newParsedCode = parse(code, { codeStyle: this.theme() });
        this.currentLineCount = this.getLineCountOfTokenArray(currentParsedCode);
        this.newLineCount = this.getLineCountOfTokenArray(newParsedCode);
        const autoWidth = this.width.isInitial();
        const autoHeight = this.height.isInitial();
        const fromSize = this.size();
        const toSize = this.getTokensSize(newParsedCode);
        const beginning = 0.2;
        const ending = 0.8;
        this.codeProgress(0);
        this.diffed = diff(this.code(), code, { codeStyle: this.theme() });
        yield* tween(time, value => {
            const progress = timingFunction(value);
            const remapped = clampRemap(beginning, ending, 0, 1, progress);
            this.codeProgress(progress);
            if (autoWidth) {
                this.width(easeInOutSine(remapped, fromSize.x, toSize.x));
            }
            if (autoHeight) {
                this.height(easeInOutSine(remapped, fromSize.y, toSize.y));
            }
        }, () => {
            this.codeProgress(null);
            this.diffed = null;
            if (autoWidth) {
                this.width.reset();
            }
            if (autoHeight) {
                this.height.reset();
            }
            this.code(code);
        });
    }
    draw(context) {
        if (!CodeBlock_1.initialized())
            return;
        this.requestFontUpdate();
        this.applyStyle(context);
        context.font = this.styles.font;
        context.textBaseline = 'top';
        const lh = parseFloat(this.styles.lineHeight);
        const w = context.measureText('X').width;
        const size = this.computedSize();
        const progress = this.codeProgress();
        const unselectedOpacity = this.unselectedOpacity();
        const globalAlpha = context.globalAlpha;
        const getSelectionAlpha = (x, y) => map(unselectedOpacity, 1, this.selectionStrength(x, y));
        const drawToken = (code, position, alpha = 1) => {
            for (let i = 0; i < code.length; i++) {
                const char = code.charAt(i);
                if (char === '\n') {
                    position.y++;
                    position.x = 0;
                    continue;
                }
                context.globalAlpha =
                    globalAlpha * alpha * getSelectionAlpha(position.x, position.y);
                context.fillText(char, position.x * w, position.y * lh);
                position.x++;
            }
        };
        context.translate(size.x / -2, size.y / -2);
        if (progress == null) {
            const parsed = this.parsed();
            const position = { x: 0, y: 0 };
            for (const token of parsed) {
                context.save();
                context.fillStyle = token.color ?? '#c9d1d9';
                drawToken(token.code, position);
                context.restore();
            }
        }
        else {
            const diffed = this.diffed;
            const beginning = 0.2;
            const ending = 0.8;
            const overlap = 0.15;
            for (const token of diffed) {
                context.save();
                context.fillStyle = token.color ?? '#c9d1d9';
                if (token.morph === 'delete') {
                    drawToken(token.code, { x: token.from[0], y: token.from[1] }, clampRemap(0, beginning + overlap, 1, 0, progress));
                }
                else if (token.morph === 'create') {
                    drawToken(token.code, { x: token.to[0], y: token.to[1] }, clampRemap(ending - overlap, 1, 0, 1, progress));
                }
                else if (token.morph === 'retain') {
                    const remapped = clampRemap(beginning, ending, 0, 1, progress);
                    const x = easeInOutSine(remapped, token.from[0], token.to[0]);
                    const y = easeInOutSine(remapped, token.from[1], token.to[1]);
                    const point = remapped > 0.5 ? token.to : token.from;
                    let offsetX = 0;
                    let offsetY = 0;
                    for (let i = 0; i < token.code.length; i++) {
                        const char = token.code.charAt(i);
                        if (char === '\n') {
                            offsetY++;
                            offsetX = 0;
                            continue;
                        }
                        context.globalAlpha =
                            globalAlpha *
                                getSelectionAlpha(point[0] + offsetX, point[1] + offsetY);
                        context.fillText(char, (x + offsetX) * w, (y + offsetY) * lh);
                        offsetX++;
                    }
                }
                else {
                    useLogger().warn({
                        message: 'Invalid token',
                        object: token,
                    });
                }
                context.restore();
            }
        }
    }
    selectionStrength(x, y) {
        const selection = this.selection();
        const selectionProgress = this.selectionProgress();
        const isSelected = CodeBlock_1.selectionStrength(selection, x, y);
        if (selectionProgress === null || this.oldSelection === null) {
            return isSelected ? 1 : 0;
        }
        const wasSelected = CodeBlock_1.selectionStrength(this.oldSelection, x, y);
        if (isSelected === wasSelected) {
            return isSelected;
        }
        return map(wasSelected, isSelected, selectionProgress);
    }
    static selectionStrength(selection, x, y) {
        return selection.length > 0 &&
            !!selection.find(([[startLine, startColumn], [endLine, endColumn]]) => {
                return (((y === startLine && x >= startColumn) || y > startLine) &&
                    ((y === endLine && x < endColumn) || y < endLine));
            })
            ? 1
            : 0;
    }
};
CodeBlock.initialized = createComputedAsync(() => ready().then(() => true), false);
__decorate([
    initial('tsx'),
    signal()
], CodeBlock.prototype, "language", void 0);
__decorate([
    initial(''),
    parser(function (value) {
        return typeof value === 'string'
            ? {
                language: this.language(),
                spans: [value],
                nodes: [],
            }
            : value;
    }),
    signal()
], CodeBlock.prototype, "code", void 0);
__decorate([
    initial(undefined),
    signal()
], CodeBlock.prototype, "theme", void 0);
__decorate([
    initial(lines(0, Infinity)),
    signal()
], CodeBlock.prototype, "selection", void 0);
__decorate([
    initial(0.32),
    signal()
], CodeBlock.prototype, "unselectedOpacity", void 0);
__decorate([
    computed()
], CodeBlock.prototype, "lineCount", null);
__decorate([
    computed()
], CodeBlock.prototype, "parsed", null);
__decorate([
    computed()
], CodeBlock.prototype, "characterSize", null);
__decorate([
    threadable()
], CodeBlock.prototype, "tweenCode", null);
CodeBlock = CodeBlock_1 = __decorate([
    nodeName('CodeBlock')
], CodeBlock);
export { CodeBlock };
function isCodeModification(value) {
    return (value &&
        typeof value === 'object' &&
        value.from !== undefined &&
        value.to !== undefined);
}
/**
 * Create a code modification that inserts a piece of code.
 *
 * @remarks
 * Should be used in conjunction with {@link CodeBlock.edit}.
 *
 * @param content - The code to insert.
 */
export function insert(content) {
    return {
        from: '',
        to: content,
    };
}
/**
 * Create a code modification that removes a piece of code.
 *
 * @remarks
 * Should be used in conjunction with {@link CodeBlock.edit}.
 *
 * @param content - The code to remove.
 */
export function remove(content) {
    return {
        from: content,
        to: '',
    };
}
/**
 * Create a code modification that changes one piece of code into another.
 *
 * @remarks
 * Should be used in conjunction with {@link CodeBlock.edit}.
 *
 * @param from - The code to change from.
 * @param to - The code to change to.
 */
export function edit(from, to) {
    return { from, to };
}
/**
 * Create a selection range that highlights the given lines.
 *
 * @param from - The line from which the selection starts.
 * @param to - The line at which the selection ends. If omitted, the selection
 *             will cover only one line.
 */
export function lines(from, to) {
    return [
        [
            [from, 0],
            [to ?? from, Infinity],
        ],
    ];
}
/**
 * Create a selection range that highlights the given word.
 *
 * @param line - The line at which the word appears.
 * @param from - The column at which the word starts.
 * @param length - The length of the word. If omitted, the selection will cover
 *                 the rest of the line.
 */
export function word(line, from, length) {
    return [
        [
            [line, from],
            [line, from + (length ?? Infinity)],
        ],
    ];
}
/**
 * Create a custom selection range.
 *
 * @param startLine - The line at which the selection starts.
 * @param startColumn - The column at which the selection starts.
 * @param endLine - The line at which the selection ends.
 * @param endColumn - The column at which the selection ends.
 */
export function range(startLine, startColumn, endLine, endColumn) {
    return [
        [
            [startLine, startColumn],
            [endLine, endColumn],
        ],
    ];
}
