// @motion-canvas/2d/lib/components/Knot.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { Vector2, } from '@motion-canvas/core';
import { cloneable, compound, computed, initial, parser, signal, wrapper, } from '../decorators';
import { Node } from './Node';
/**
 * A node representing a knot of a {@link Spline}.
 */
export class Knot extends Node {
    get startHandleAuto() {
        return this.auto.startHandle;
    }
    get endHandleAuto() {
        return this.auto.endHandle;
    }
    constructor(props) {
        super(props.startHandle === undefined && props.endHandle === undefined
            ? { auto: 1, ...props }
            : props);
    }
    points() {
        const hasExplicitHandles = !this.startHandle.isInitial() || !this.endHandle.isInitial();
        const startHandle = hasExplicitHandles ? this.startHandle() : Vector2.zero;
        const endHandle = hasExplicitHandles ? this.endHandle() : Vector2.zero;
        return {
            position: this.position(),
            startHandle: startHandle.transformAsPoint(this.localToParent()),
            endHandle: endHandle.transformAsPoint(this.localToParent()),
            auto: { start: this.startHandleAuto(), end: this.endHandleAuto() },
        };
    }
    getDefaultEndHandle() {
        return this.startHandle().flipped;
    }
    getDefaultStartHandle() {
        return this.endHandle().flipped;
    }
}
__decorate([
    wrapper(Vector2),
    signal()
], Knot.prototype, "startHandle", void 0);
__decorate([
    wrapper(Vector2),
    signal()
], Knot.prototype, "endHandle", void 0);
__decorate([
    cloneable(false),
    initial(() => ({ startHandle: 0, endHandle: 0 })),
    parser((value) => {
        if (typeof value === 'object' && !Array.isArray(value)) {
            return value;
        }
        if (typeof value === 'number') {
            value = [value, value];
        }
        return { startHandle: value[0], endHandle: value[1] };
    }),
    compound({ startHandle: 'startHandleAuto', endHandle: 'endHandleAuto' })
], Knot.prototype, "auto", void 0);
__decorate([
    computed()
], Knot.prototype, "points", null);
