// @motion-canvas/2d/lib/components/Latex.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { lazy, threadable, useLogger, Vector2, } from '@motion-canvas/core';
import { liteAdaptor } from 'mathjax-full/js/adaptors/liteAdaptor';
import { RegisterHTMLHandler } from 'mathjax-full/js/handlers/html';
import { TeX } from 'mathjax-full/js/input/tex';
import { AllPackages } from 'mathjax-full/js/input/tex/AllPackages';
import { mathjax } from 'mathjax-full/js/mathjax';
import { SVG } from 'mathjax-full/js/output/svg';
import { computed, initial, parser, signal } from '../decorators';
import { Node } from './Node';
import { SVG as SVGNode, } from './SVG';
const Adaptor = liteAdaptor();
RegisterHTMLHandler(Adaptor);
const JaxDocument = mathjax.document('', {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    InputJax: new TeX({ packages: AllPackages }),
    // eslint-disable-next-line @typescript-eslint/naming-convention
    OutputJax: new SVG({ fontCache: 'local' }),
});
/**
 * A node for animating equations with LaTeX.
 *
 * @preview
 * ```tsx editor
 * import {Latex, makeScene2D} from '@motion-canvas/2d';
 * import {createRef, waitFor} from '@motion-canvas/core';
 *
 * export default makeScene2D(function* (view) {
 *   const tex = createRef<Latex>();
 *   view.add(<Latex ref={tex} tex="{{y=}}{{a}}{{x^2}}" fill="white" />);
 *
 *   yield* waitFor(0.2);
 *   yield* tex().tex('{{y=}}{{a}}{{x^2}} + {{bx}}', 1);
 *   yield* waitFor(0.2);
 *   yield* tex().tex(
 *     '{{y=}}{{\\left(}}{{a}}{{x^2}} + {{bx}}{{\\over 1}}{{\\right)}}',
 *     1,
 *   );
 *   yield* waitFor(0.2);
 *   yield* tex().tex('{{y=}}{{a}}{{x^2}}', 1);
 * });
 * ```
 */
export class Latex extends SVGNode {
    constructor(props) {
        super({
            fontSize: 48,
            ...props,
            svg: '',
        });
        this.svgSubTexMap = {};
        this.svg(this.latexSVG);
    }
    calculateWrapperScale(documentSize, parentSize) {
        if (parentSize.x || parentSize.y) {
            return super.calculateWrapperScale(documentSize, parentSize);
        }
        return new Vector2(this.fontSize() / Latex.containerFontSize);
    }
    latexSVG() {
        return this.texToSvg(this.tex());
    }
    getNodeCharacterId({ id }) {
        if (!id.includes('-'))
            return id;
        return id.substring(id.lastIndexOf('-') + 1);
    }
    parseSVG(svg) {
        const subTexs = this.svgSubTexMap[svg].map(sub => sub.trim());
        const key = `[${subTexs.join(',')}]::${JSON.stringify(this.options())}`;
        const cached = Latex.texNodesPool[key];
        if (cached && (cached.size.x > 0 || cached.size.y > 0)) {
            return this.buildDocument(Latex.texNodesPool[key]);
        }
        const oldSVG = SVGNode.parseSVGData(svg);
        const oldNodes = [...oldSVG.nodes];
        const newNodes = [];
        for (const sub of subTexs) {
            const subSvg = this.subTexToSVG(sub);
            const subNodes = SVGNode.parseSVGData(subSvg).nodes;
            if (subNodes.length === 0) {
                continue;
            }
            const firstId = this.getNodeCharacterId(subNodes[0]);
            const spliceIndex = oldNodes.findIndex(node => this.getNodeCharacterId(node) === firstId);
            const children = oldNodes.splice(spliceIndex, subNodes.length);
            if (children.length === 1) {
                newNodes.push({
                    ...children[0],
                    id: sub,
                });
                continue;
            }
            newNodes.push({
                id: sub,
                type: Node,
                props: {},
                children,
            });
        }
        if (oldNodes.length > 0) {
            useLogger().error({
                message: 'Matching between Latex SVG and tex parts failed',
                inspect: this.key,
            });
        }
        const newSVG = {
            size: oldSVG.size,
            nodes: newNodes,
        };
        Latex.texNodesPool[key] = newSVG;
        return this.buildDocument(newSVG);
    }
    texToSvg(subTexs) {
        const singleTex = subTexs.join('');
        const svg = this.singleTexToSVG(singleTex);
        this.svgSubTexMap[svg] = subTexs;
        return svg;
    }
    subTexToSVG(subTex) {
        let tex = subTex.trim();
        if (['\\overline', '\\sqrt', '\\sqrt{'].includes(tex) ||
            tex.endsWith('_') ||
            tex.endsWith('^') ||
            tex.endsWith('dot')) {
            tex += '{\\quad}';
        }
        if (tex === '\\substack')
            tex = '\\quad';
        const numLeft = tex.match(/\\left[()[\]|.\\]/g)?.length ?? 0;
        const numRight = tex.match(/\\right[()[\]|.\\]/g)?.length ?? 0;
        if (numLeft !== numRight) {
            tex = tex.replace(/\\left/g, '\\big').replace(/\\right/g, '\\big');
        }
        const bracesLeft = tex.match(/((?<!\\)|(?<=\\\\)){/g)?.length ?? 0;
        const bracesRight = tex.match(/((?<!\\)|(?<=\\\\))}/g)?.length ?? 0;
        if (bracesLeft < bracesRight) {
            tex = '{'.repeat(bracesRight - bracesLeft) + tex;
        }
        else if (bracesRight < bracesLeft) {
            tex += '}'.repeat(bracesLeft - bracesRight);
        }
        const hasArrayBegin = tex.includes('\\begin{array}');
        const hasArrayEnd = tex.includes('\\end{array}');
        if (hasArrayBegin !== hasArrayEnd)
            tex = '';
        return this.singleTexToSVG(tex);
    }
    singleTexToSVG(tex) {
        const src = `${tex}::${JSON.stringify(this.options())}`;
        if (Latex.svgContentsPool[src]) {
            return Latex.svgContentsPool[src];
        }
        const svg = Adaptor.innerHTML(JaxDocument.convert(tex, this.options()));
        if (svg.includes('data-mjx-error')) {
            const errors = svg.match(/data-mjx-error="(.*?)"/);
            if (errors && errors.length > 0) {
                useLogger().error(`Invalid MathJax: ${errors[1]}`);
            }
        }
        Latex.svgContentsPool[src] = svg;
        return svg;
    }
    *tweenTex(value, time, timingFunction) {
        const newSVG = this.texToSvg(this.tex.context.parse(value));
        yield* this.svg(newSVG, time, timingFunction);
        this.svg(this.latexSVG);
    }
}
Latex.svgContentsPool = {};
Latex.texNodesPool = {};
__decorate([
    initial({}),
    signal()
], Latex.prototype, "options", void 0);
__decorate([
    initial(''),
    parser(function (value) {
        const array = typeof value === 'string' ? [value] : value;
        return array
            .reduce((prev, current) => {
            prev.push(...current.split(/{{(.*?)}}/));
            return prev;
        }, [])
            .filter(sub => sub.trim().length > 0);
    }),
    signal()
], Latex.prototype, "tex", void 0);
__decorate([
    computed()
], Latex.prototype, "latexSVG", null);
__decorate([
    threadable()
], Latex.prototype, "tweenTex", null);
__decorate([
    lazy(() => {
        return parseFloat(window.getComputedStyle(SVGNode.containerElement).fontSize);
    })
], Latex, "containerFontSize", void 0);
