// @motion-canvas/2d/lib/components/Path.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { BBox, createSignal, isReactive, threadable, tween, } from '@motion-canvas/core';
import { createCurveProfileLerp } from '../curves/createCurveProfileLerp';
import { getPathProfile } from '../curves/getPathProfile';
import { computed, signal } from '../decorators';
import { drawLine, drawPivot } from '../utils';
import { Curve } from './Curve';
export class Path extends Curve {
    constructor(props) {
        super(props);
        this.currentProfile = createSignal(null);
        this.canHaveSubpath = true;
    }
    profile() {
        return this.currentProfile() ?? getPathProfile(this.data());
    }
    childrenBBox() {
        const points = this.profile().segments.flatMap(segment => segment.points);
        return BBox.fromPoints(...points);
    }
    lineWidthCoefficient() {
        const join = this.lineJoin();
        let coefficient = super.lineWidthCoefficient();
        if (join === 'miter') {
            const { minSin } = this.profile();
            if (minSin > 0) {
                coefficient = Math.max(coefficient, 0.5 / minSin);
            }
        }
        return coefficient;
    }
    processSubpath(path, startPoint, endPoint) {
        if (startPoint && endPoint && startPoint.equals(endPoint)) {
            path.closePath();
        }
    }
    *tweenData(newPath, time, timingFunction) {
        const fromProfile = this.profile();
        const toProfile = getPathProfile(isReactive(newPath) ? newPath() : newPath);
        const interpolator = createCurveProfileLerp(fromProfile, toProfile);
        this.currentProfile(fromProfile);
        yield* tween(time, value => {
            const progress = timingFunction(value);
            this.currentProfile(interpolator(progress));
        }, () => {
            this.currentProfile(null);
            this.data(newPath);
        });
    }
    drawOverlay(context, matrix) {
        const box = this.childrenBBox().transformCorners(matrix);
        const size = this.computedSize();
        const offset = size.mul(this.offset()).scale(0.5).transformAsPoint(matrix);
        const segments = this.profile().segments;
        context.lineWidth = 1;
        context.strokeStyle = 'white';
        context.fillStyle = 'white';
        context.save();
        context.setTransform(matrix);
        let endPoint = null;
        let path = new Path2D();
        for (const segment of segments) {
            if (endPoint && !segment.getPoint(0).position.equals(endPoint)) {
                context.stroke(path);
                path = new Path2D();
                endPoint = null;
            }
            const [, end] = segment.draw(path, 0, 1, endPoint == null);
            endPoint = end.position;
        }
        context.stroke(path);
        context.restore();
        context.beginPath();
        drawPivot(context, offset);
        context.stroke();
        context.beginPath();
        drawLine(context, box);
        context.closePath();
        context.stroke();
    }
}
__decorate([
    signal()
], Path.prototype, "data", void 0);
__decorate([
    computed()
], Path.prototype, "profile", null);
__decorate([
    threadable()
], Path.prototype, "tweenData", null);
