// @motion-canvas/2d/lib/components/Polygon.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { BBox, Vector2, } from '@motion-canvas/core';
import { getPolylineProfile } from '../curves';
import { computed, initial, signal } from '../decorators';
import { drawPolygon } from '../utils';
import { Curve } from './Curve';
/**
 * A node for drawing regular polygons.
 *
 * @remarks
 * This node can be used to render shapes such as: triangle, pentagon,
 * hexagon and more.
 *
 * Note that the polygon is inscribed in a circle defined by the height
 * and width. If height and width are unequal, the polygon is inscribed
 * in the resulting ellipse.
 *
 * Since the polygon is inscribed in the circle, the actual displayed
 * height and width may differ somewhat from the bounding rectangle. This
 * will be particularly noticeable if the number of sides is low, e.g. for a
 * triangle.
 *
 * @preview
 * ```tsx editor
 * // snippet Polygon
 * import {makeScene2D, Polygon} from '@motion-canvas/2d';
 * import {createRef} from '@motion-canvas/core';
 *
 * export default makeScene2D(function* (view) {
 *   const ref = createRef<Polygon>();
 *   view.add(
 *     <Polygon
 *       ref={ref}
 *       sides={6}
 *       size={160}
 *       fill={'lightseagreen'}
 *     />
 *   );
 *
 *   yield* ref().sides(3, 2).to(6, 2);
 * });
 *
 * // snippet Pentagon outline
 * import {makeScene2D, Polygon} from '@motion-canvas/2d';
 *
 * export default makeScene2D(function* (view) {
 *   view.add(
 *     <Polygon
 *       sides={5}
 *       size={160}
 *       radius={30}
 *       stroke={'lightblue'}
 *       lineWidth={8}
 *     />
 *   );
 * });
 *
 * // snippet Accessing vertex data
 * import {Circle, Polygon, makeScene2D} from '@motion-canvas/2d';
 * import {createRef, range} from '@motion-canvas/core';
 *
 * export default makeScene2D(function* (view) {
 *   const polygon = createRef<Polygon>();
 *   view.add(
 *     <Polygon ref={polygon} sides={3} lineWidth={4} stroke={'white'} size={160}>
 *       {range(6).map(index => (
 *         <Circle
 *           fill={'white'}
 *           size={20}
 *           position={() => polygon().vertex(index)}
 *           opacity={() => polygon().vertexCompletion(index)}
 *         />
 *       ))}
 *     </Polygon>,
 *   );
 *
 *   yield* polygon().sides(6, 2).wait(0.5).back(2);
 * });
 * ```
 */
export class Polygon extends Curve {
    constructor(props) {
        super(props);
    }
    /**
     * Get the position of the nth vertex in the local space of this polygon.
     *
     * @param index - The index of the vertex.
     */
    vertex(index) {
        const size = this.computedSize().scale(0.5);
        const theta = (index * 2 * Math.PI) / this.sides();
        const direction = Vector2.fromRadians(theta).perpendicular;
        return direction.mul(size);
    }
    /**
     * Get the completion of the nth vertex.
     *
     * @remarks
     * The completion is a value between `0` and `1` that describes how the given
     * vertex partakes in the polygon.
     *
     * For integer values of {@link sides}, the completion is simply `1` for
     * each index making up the polygon and `0` for any other index. If `sides`
     * includes a fraction, the last index of the polygon will have a completion
     * equal to said fraction.
     *
     * Check out the {@link Polygon | Accessing vertex data} example for a
     * demonstration.
     *
     * @param index - The index of the vertex.
     */
    vertexCompletion(index) {
        const sides = this.sides();
        if (index < 0 || index > sides) {
            return 0;
        }
        if (index < sides - 1) {
            return 1;
        }
        return sides - index;
    }
    profile() {
        const sides = this.sides();
        const radius = this.radius();
        const points = [];
        const size = this.computedSize().scale(0.5);
        for (let i = 0; i < sides; i++) {
            const theta = (i * 2 * Math.PI) / sides;
            const direction = Vector2.fromRadians(theta).perpendicular;
            points.push(direction.mul(size));
        }
        return getPolylineProfile(points, radius, true);
    }
    desiredSize() {
        return {
            x: this.width.context.getter(),
            y: this.height.context.getter(),
        };
    }
    offsetComputedLayout(box) {
        return box;
    }
    childrenBBox() {
        return BBox.fromSizeCentered(this.computedSize());
    }
    requiresProfile() {
        return super.requiresProfile() || this.radius() > 0;
    }
    getPath() {
        if (this.requiresProfile()) {
            return this.curveDrawingInfo().path;
        }
        return this.createPath();
    }
    getRipplePath() {
        return this.createPath(this.rippleSize());
    }
    createPath(expand = 0) {
        const path = new Path2D();
        const sides = this.sides();
        const box = BBox.fromSizeCentered(this.size()).expand(expand);
        drawPolygon(path, box, sides);
        return path;
    }
}
__decorate([
    initial(6),
    signal()
], Polygon.prototype, "sides", void 0);
__decorate([
    initial(0),
    signal()
], Polygon.prototype, "radius", void 0);
__decorate([
    computed()
], Polygon.prototype, "profile", null);
