// @motion-canvas/2d/lib/components/Ray.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { BBox, } from '@motion-canvas/core';
import { LineSegment } from '../curves';
import { nodeName, vector2Signal } from '../decorators';
import { arc, drawLine, drawPivot } from '../utils';
import { Curve } from './Curve';
/**
 * A node for drawing an individual line segment.
 *
 * @preview
 * ```tsx editor
 * import {makeScene2D} from '@motion-canvas/2d';
 * import {Ray} from '@motion-canvas/2d';
 * import {createRef} from '@motion-canvas/core';
 *
 * export default makeScene2D(function* (view) {
 *   const ray = createRef<Ray>();
 *
 *   view.add(
 *     <Ray
 *       ref={ray}
 *       lineWidth={8}
 *       endArrow
 *       stroke={'lightseagreen'}
 *       fromX={-200}
 *       toX={200}
 *     />,
 *   );
 *
 *   yield* ray().start(1, 1);
 *   yield* ray().start(0).end(0).start(1, 1);
 * });
 * ```
 */
let Ray = class Ray extends Curve {
    constructor(props) {
        super(props);
    }
    childrenBBox() {
        return BBox.fromPoints(this.from(), this.to());
    }
    profile() {
        const segment = new LineSegment(this.from(), this.to());
        return {
            arcLength: segment.arcLength,
            minSin: 1,
            segments: [segment],
        };
    }
    drawOverlay(context, matrix) {
        const box = this.childrenBBox().transformCorners(matrix);
        const size = this.computedSize();
        const offset = size.mul(this.offset()).scale(0.5).transformAsPoint(matrix);
        const from = this.from().transformAsPoint(matrix);
        const to = this.to().transformAsPoint(matrix);
        context.fillStyle = 'white';
        context.strokeStyle = 'black';
        context.lineWidth = 1;
        context.beginPath();
        arc(context, from, 4);
        context.fill();
        context.stroke();
        context.beginPath();
        arc(context, to, 4);
        context.fill();
        context.stroke();
        context.strokeStyle = 'white';
        context.beginPath();
        drawLine(context, [from, to]);
        context.stroke();
        context.beginPath();
        drawPivot(context, offset);
        context.stroke();
        context.beginPath();
        drawLine(context, box);
        context.closePath();
        context.stroke();
    }
};
__decorate([
    vector2Signal('from')
], Ray.prototype, "from", void 0);
__decorate([
    vector2Signal('to')
], Ray.prototype, "to", void 0);
Ray = __decorate([
    nodeName('Ray')
], Ray);
export { Ray };
