// @motion-canvas/2d/lib/components/Rect.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { BBox, } from '@motion-canvas/core';
import { getRectProfile } from '../curves/getRectProfile';
import { computed, initial, nodeName, signal } from '../decorators';
import { spacingSignal } from '../decorators/spacingSignal';
import { drawRoundRect } from '../utils';
import { Curve } from './Curve';
let Rect = class Rect extends Curve {
    constructor(props) {
        super(props);
    }
    profile() {
        return getRectProfile(this.childrenBBox(), this.radius(), this.smoothCorners(), this.cornerSharpness());
    }
    desiredSize() {
        return {
            x: this.width.context.getter(),
            y: this.height.context.getter(),
        };
    }
    offsetComputedLayout(box) {
        return box;
    }
    childrenBBox() {
        return BBox.fromSizeCentered(this.computedSize());
    }
    getPath() {
        if (this.requiresProfile()) {
            return this.curveDrawingInfo().path;
        }
        const path = new Path2D();
        const radius = this.radius();
        const smoothCorners = this.smoothCorners();
        const cornerSharpness = this.cornerSharpness();
        const box = BBox.fromSizeCentered(this.size());
        drawRoundRect(path, box, radius, smoothCorners, cornerSharpness);
        return path;
    }
    getCacheBBox() {
        return super.getCacheBBox().expand(this.rippleSize());
    }
    getRipplePath() {
        const path = new Path2D();
        const rippleSize = this.rippleSize();
        const radius = this.radius().addScalar(rippleSize);
        const smoothCorners = this.smoothCorners();
        const cornerSharpness = this.cornerSharpness();
        const box = BBox.fromSizeCentered(this.size()).expand(rippleSize);
        drawRoundRect(path, box, radius, smoothCorners, cornerSharpness);
        return path;
    }
};
__decorate([
    spacingSignal('radius')
], Rect.prototype, "radius", void 0);
__decorate([
    initial(false),
    signal()
], Rect.prototype, "smoothCorners", void 0);
__decorate([
    initial(0.6),
    signal()
], Rect.prototype, "cornerSharpness", void 0);
__decorate([
    computed()
], Rect.prototype, "profile", null);
Rect = __decorate([
    nodeName('Rect')
], Rect);
export { Rect };
