// @motion-canvas/2d/lib/components/Video.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var Video_1;
import { BBox, DependencyContext, PlaybackState, clamp, isReactive, useLogger, useThread, } from '@motion-canvas/core';
import { computed, initial, nodeName, signal } from '../decorators';
import { drawImage } from '../utils';
import { Rect } from './Rect';
let Video = Video_1 = class Video extends Rect {
    constructor({ play, ...props }) {
        super(props);
        this.lastTime = -1;
        if (play) {
            this.play();
        }
    }
    isPlaying() {
        return this.playing();
    }
    getCurrentTime() {
        return this.clampTime(this.time());
    }
    getDuration() {
        return this.video().duration;
    }
    desiredSize() {
        const custom = super.desiredSize();
        if (custom.x === null && custom.y === null) {
            const image = this.video();
            return {
                x: image.videoWidth,
                y: image.videoHeight,
            };
        }
        return custom;
    }
    completion() {
        return this.clampTime(this.time()) / this.video().duration;
    }
    video() {
        const src = this.src();
        const key = `${this.key}/${src}`;
        let video = Video_1.pool[key];
        if (!video) {
            video = document.createElement('video');
            video.src = src;
            Video_1.pool[key] = video;
        }
        if (video.readyState < 2) {
            DependencyContext.collectPromise(new Promise(resolve => {
                const listener = () => {
                    resolve();
                    video.removeEventListener('canplay', listener);
                };
                video.addEventListener('canplay', listener);
            }));
        }
        return video;
    }
    seekedVideo() {
        const video = this.video();
        const time = this.clampTime(this.time());
        video.playbackRate = this.playbackRate();
        if (!video.paused) {
            video.pause();
        }
        if (this.lastTime === time) {
            return video;
        }
        this.setCurrentTime(time);
        return video;
    }
    fastSeekedVideo() {
        const video = this.video();
        const time = this.clampTime(this.time());
        video.playbackRate = this.playbackRate();
        if (this.lastTime === time) {
            return video;
        }
        const playing = this.playing() && time < video.duration && video.playbackRate > 0;
        if (playing) {
            if (video.paused) {
                DependencyContext.collectPromise(video.play());
            }
        }
        else {
            if (!video.paused) {
                video.pause();
            }
        }
        if (Math.abs(video.currentTime - time) > 0.2) {
            this.setCurrentTime(time);
        }
        else if (!playing) {
            video.currentTime = time;
        }
        return video;
    }
    draw(context) {
        this.drawShape(context);
        const alpha = this.alpha();
        if (alpha > 0) {
            const playbackState = this.view().playbackState();
            const video = playbackState === PlaybackState.Playing ||
                playbackState === PlaybackState.Presenting
                ? this.fastSeekedVideo()
                : this.seekedVideo();
            const box = BBox.fromSizeCentered(this.computedSize());
            context.save();
            context.clip(this.getPath());
            if (alpha < 1) {
                context.globalAlpha *= alpha;
            }
            context.imageSmoothingEnabled = this.smoothing();
            drawImage(context, video, box);
            context.restore();
        }
        if (this.clip()) {
            context.clip(this.getPath());
        }
        this.drawChildren(context);
    }
    applyFlex() {
        super.applyFlex();
        const video = this.video();
        this.element.style.aspectRatio = (this.ratio() ?? video.videoWidth / video.videoHeight).toString();
    }
    setCurrentTime(value) {
        const video = this.video();
        if (video.readyState < 2)
            return;
        video.currentTime = value;
        this.lastTime = value;
        if (video.seeking) {
            DependencyContext.collectPromise(new Promise(resolve => {
                const listener = () => {
                    resolve();
                    video.removeEventListener('seeked', listener);
                };
                video.addEventListener('seeked', listener);
            }));
        }
    }
    setPlaybackRate(playbackRate) {
        let value;
        if (isReactive(playbackRate)) {
            value = playbackRate();
            useLogger().warn({
                message: 'Invalid value set as the playback rate',
                remarks: "<p>The <code>playbackRate</code> of a <code>Video</code> cannot be reactive.</p>\n<p>Make sure to use a concrete value and not a function:</p>\n<pre class=\"wrong\"><code class=\"language-ts\">video.<span class=\"hljs-title function_\">playbackRate</span>(<span class=\"hljs-function\">() =&gt;</span> <span class=\"hljs-number\">7</span>);</code></pre><pre class=\"correct\"><code class=\"language-ts\">video.<span class=\"hljs-title function_\">playbackRate</span>(<span class=\"hljs-number\">7</span>);</code></pre><p>If you&#39;re using a signal, extract its value before passing it to the property:</p>\n<pre class=\"wrong\"><code class=\"language-ts\">video.<span class=\"hljs-title function_\">playbackRate</span>(mySignal);</code></pre><pre class=\"correct\"><code class=\"language-ts\">video.<span class=\"hljs-title function_\">playbackRate</span>(<span class=\"hljs-title function_\">mySignal</span>());</code></pre>",
                inspect: this.key,
                stack: new Error().stack,
            });
        }
        else {
            value = playbackRate;
        }
        this.playbackRate.context.setter(value);
        if (this.playing()) {
            if (value === 0) {
                this.pause();
            }
            else {
                const time = useThread().time;
                const start = time();
                const offset = this.time();
                this.time(() => this.clampTime(offset + (time() - start) * value));
            }
        }
    }
    play() {
        const time = useThread().time;
        const start = time();
        const offset = this.time();
        const playbackRate = this.playbackRate();
        this.playing(true);
        this.time(() => this.clampTime(offset + (time() - start) * playbackRate));
    }
    pause() {
        this.playing(false);
        this.time.save();
        this.video().pause();
    }
    seek(time) {
        const playing = this.playing();
        this.time(this.clampTime(time));
        if (playing) {
            this.play();
        }
        else {
            this.pause();
        }
    }
    clampTime(time) {
        const duration = this.video().duration;
        if (this.loop()) {
            time %= duration;
        }
        return clamp(0, duration, time);
    }
    collectAsyncResources() {
        super.collectAsyncResources();
        this.seekedVideo();
    }
};
Video.pool = {};
__decorate([
    signal()
], Video.prototype, "src", void 0);
__decorate([
    initial(1),
    signal()
], Video.prototype, "alpha", void 0);
__decorate([
    initial(true),
    signal()
], Video.prototype, "smoothing", void 0);
__decorate([
    initial(false),
    signal()
], Video.prototype, "loop", void 0);
__decorate([
    initial(1),
    signal()
], Video.prototype, "playbackRate", void 0);
__decorate([
    initial(0),
    signal()
], Video.prototype, "time", void 0);
__decorate([
    initial(false),
    signal()
], Video.prototype, "playing", void 0);
__decorate([
    computed()
], Video.prototype, "completion", null);
__decorate([
    computed()
], Video.prototype, "video", null);
__decorate([
    computed()
], Video.prototype, "seekedVideo", null);
__decorate([
    computed()
], Video.prototype, "fastSeekedVideo", null);
Video = Video_1 = __decorate([
    nodeName('Video')
], Video);
export { Video };
