// @motion-canvas/core/lib/meta/ExporterMetaFile.js
import { ValueDispatcher } from '../events';
import { EnumMetaField } from './EnumMetaField';
import { MetaField } from './MetaField';
/**
 * Represents the exporter configuration.
 */
export class ExporterMetaField extends MetaField {
    /**
     * Triggered when the nested fields change.
     *
     * @eventProperty
     */
    get onFieldsChanged() {
        return this.fields.subscribable;
    }
    get options() {
        return this.optionFields[this.current];
    }
    constructor(name, project, current = 0) {
        const exporters = project.plugins.flatMap(plugin => plugin.exporters?.(project) ?? []);
        const optionFields = exporters.map(exporter => exporter.meta(project));
        const exporterField = new EnumMetaField('exporter', exporters.map(exporter => ({
            value: exporter.id,
            text: exporter.displayName,
        })), exporters[current]?.id);
        super(name, {
            name: exporterField.get(),
            options: optionFields[current]?.get(),
        });
        this.current = current;
        this.type = Object;
        this.handleChange = () => {
            const value = this.exporterField.get();
            const index = Math.max(this.exporters.findIndex(exporter => exporter.id === value), 0);
            if (this.current !== index) {
                this.options?.onChanged.unsubscribe(this.handleChange);
                this.current = index;
                this.options?.onChanged.subscribe(this.handleChange, false);
                this.fields.current = this.options
                    ? [this.exporterField, this.options]
                    : [this.exporterField];
            }
            this.value.current = {
                name: this.exporterField.get(),
                options: this.options?.get() ?? null,
            };
        };
        this.exporters = exporters;
        this.exporterField = exporterField;
        this.exporterField.onChanged.subscribe(this.handleChange, false);
        this.exporterField.disable(optionFields.length < 2).space();
        this.optionFields = optionFields;
        this.fields = new ValueDispatcher([this.exporterField]);
        if (this.options) {
            this.options.onChanged.subscribe(this.handleChange, false);
            this.fields.current = [this.exporterField, this.options];
        }
    }
    set(value) {
        this.exporterField.set(value.name);
        this.options?.set(value.options ?? {});
    }
    serialize() {
        return {
            name: this.exporterField.serialize(),
            options: this.options?.serialize() ?? null,
        };
    }
    clone() {
        return new this.constructor(this.name, this.exporters, this.current);
    }
}
