// @motion-canvas/core/lib/meta/ObjectMetaField.js
import { ValueDispatcher } from '../events';
import { MetaField } from './MetaField';
class ObjectMetaFieldInternal extends MetaField {
    /**
     * Triggered when the nested fields change.
     *
     * @eventProperty
     */
    get onFieldsChanged() {
        return this.event.subscribable;
    }
    constructor(name, fields) {
        const map = new Map(Object.entries(fields));
        super(name, Object.fromEntries(Array.from(map, ([name, field]) => [name, field.get()])));
        this.type = Object;
        this.ignoreChange = false;
        this.customFields = {};
        this.handleChange = () => {
            if (this.ignoreChange)
                return;
            this.value.current = {
                ...this.transform('get'),
                ...this.customFields,
            };
        };
        this.event = new ValueDispatcher([...map.values()]);
        this.fields = map;
        for (const [key, field] of this.fields) {
            Object.defineProperty(this, key, { value: field });
            field.onChanged.subscribe(this.handleChange);
        }
    }
    set(value) {
        this.ignoreChange = true;
        for (const [key, fieldValue] of Object.entries(value)) {
            const field = this.fields.get(key);
            if (field) {
                field.set(fieldValue);
            }
            else {
                this.customFields[key] = fieldValue;
            }
        }
        this.ignoreChange = false;
        this.handleChange();
    }
    serialize() {
        return {
            ...this.transform('serialize'),
            ...this.customFields,
        };
    }
    clone() {
        const cloned = new this.constructor(this.name, this.transform('clone'));
        cloned.set(structuredClone(this.customFields));
        return cloned;
    }
    transform(fn) {
        const transformed = Object.fromEntries(Array.from(this.fields, ([name, field]) => [name, field[fn]()]));
        return transformed;
    }
}
/**
 * Represents an object with nested meta-fields.
 */
export const ObjectMetaField = ObjectMetaFieldInternal;
