// @motion-canvas/core/lib/tweening/interpolationFunctions.js
import { Vector2 } from '../types';
import { useLogger } from '../utils';
export function textLerp(fromString, toString, value) {
    const from = [...fromString];
    const to = [...toString];
    // left to right
    if (to.length >= from.length) {
        const current = Math.floor(to.length * value);
        const currentLength = Math.floor(map(from.length - 1, to.length, value));
        let text = '';
        for (let i = 0; i < to.length; i++) {
            if (i < current) {
                text += to[i];
            }
            else if (from[i] || i <= currentLength) {
                text += from[i] ?? to[i];
            }
        }
        return text;
    }
    // right to left
    else {
        const current = Math.round(from.length * (1 - value));
        const currentLength = Math.floor(map(from.length + 1, to.length, value));
        const text = [];
        for (let i = from.length - 1; i >= 0; i--) {
            if (i < current) {
                text.unshift(from[i]);
            }
            else if (to[i] || i < currentLength) {
                text.unshift(to[i] ?? from[i]);
            }
        }
        return text.join('');
    }
}
export function deepLerp(from, to, value, suppressWarnings = false) {
    if (value === 0)
        return from;
    if (value === 1)
        return to;
    if (from == null || to == null) {
        if (!suppressWarnings) {
            useLogger().warn(`Attempting to lerp ${from} -> ${to} may result in unexpected behavior.`);
        }
        return undefined;
    }
    if (typeof from === 'number' && typeof to === 'number') {
        return map(from, to, value);
    }
    if (typeof from === 'string' && typeof to === 'string') {
        return textLerp(from, to, value);
    }
    if (typeof from === 'boolean' && typeof to === 'boolean') {
        return value < 0.5 ? from : to;
    }
    if ('lerp' in from) {
        return from.lerp(to, value);
    }
    if (from && to && typeof from === 'object' && typeof to === 'object') {
        if (Array.isArray(from) && Array.isArray(to)) {
            if (from.length === to.length) {
                return from.map((f, i) => deepLerp(f, to[i], value));
            }
        }
        else {
            let toObject = false;
            if (!(from instanceof Map) && !(to instanceof Map)) {
                toObject = true;
                from = new Map(Object.entries(from));
                to = new Map(Object.entries(to));
            }
            if (from instanceof Map && to instanceof Map) {
                const result = new Map();
                for (const key of new Set([...from.keys(), ...to.keys()])) {
                    const inter = deepLerp(from.get(key), to.get(key), value, true);
                    if (inter !== undefined)
                        result.set(key, inter);
                }
                return toObject ? Object.fromEntries(result) : result;
            }
        }
    }
    // fallback with an immediate jump to the new value
    return to;
}
export function boolLerp(from, to, value) {
    return value < 0.5 ? from : to;
}
export function map(from, to, value) {
    return from + (to - from) * value;
}
export function remap(fromIn, toIn, fromOut, toOut, value) {
    return fromOut + ((value - fromIn) * (toOut - fromOut)) / (toIn - fromIn);
}
export function clamp(min, max, value) {
    return value < min ? min : value > max ? max : value;
}
export function clampRemap(fromIn, toIn, fromOut, toOut, value) {
    const remappedValue = remap(fromIn, toIn, fromOut, toOut, value);
    if (fromOut > toOut)
        [fromOut, toOut] = [toOut, fromOut];
    return clamp(fromOut, toOut, remappedValue);
}
export function arcLerp(value, reverse, ratio) {
    let flip = reverse;
    if (ratio > 1) {
        ratio = 1 / ratio;
    }
    else {
        flip = !flip;
    }
    const normalized = flip
        ? Math.acos(clamp(-1, 1, 1 - value))
        : Math.asin(value);
    const radians = map(normalized, map(0, Math.PI / 2, value), ratio);
    let xValue = Math.sin(radians);
    let yValue = 1 - Math.cos(radians);
    if (reverse) {
        [xValue, yValue] = [yValue, xValue];
    }
    return new Vector2(xValue, yValue);
}
