// @motion-canvas/core/lib/scenes/timeEvents/ReadOnlyTimeEvents.js
import { ValueDispatcher } from '../../events';
/**
 * Manages time events during rendering and presentation.
 */
export class ReadOnlyTimeEvents {
    get onChanged() {
        return this.events.subscribable;
    }
    constructor(scene) {
        this.scene = scene;
        this.events = new ValueDispatcher([]);
        this.lookup = new Map();
        /**
         * Called when the parent scene gets reloaded.
         */
        this.handleReload = () => {
            this.lookup.clear();
        };
        scene.onReloaded.subscribe(this.handleReload);
    }
    set() {
        // do nothing
    }
    register(name, initialTime) {
        let duration = this.lookup.get(name);
        if (duration === undefined) {
            const event = this.scene.meta.timeEvents
                .get()
                .find(event => event.name === name);
            duration = event ? event.targetTime - initialTime : 0;
            this.lookup.set(name, duration);
        }
        return duration;
    }
}
